---
title: "03_Deconvolution_CTVT_geneCorrelations"
author: "IMEV2"
date: "`r Sys.Date()`"
output: html_document
---

# 0 Environment
## 0.1 Packages
### 0.1.1 Installs
```{r echo=F, message=F, warning=FALSE}
# if (!requireNamespace("devtools", quietly = TRUE)) {
#    install.packages("devtools")
# }
# devtools::install_github("Danko-Lab/BayesPrism/BayesPrism")
# install.packages("paletteer")
# Developmental version
# devtools::install_github("awhstin/awtools")
# install.packages("ggsci")
# install.packages("ggnewscale")


```
### 0.1.1 Library
```{r echo=F, message=F, warning=FALSE}
set.seed(123)
library(tidyverse)
library(openxlsx)
library(SingleCellExperiment)
library(ggrepel)
setClassUnion("ExpData", c("matrix", "SummarizedExperiment"))
# setClassUnion("characterORNULL", c("character", "NULL"))
# setClassUnion("DataTableORNULL", c("DataTable", "NULL"))

library(glmGamPoi)
library(DESeq2)
library(paletteer)
library(scales)
library(ggnewscale)
library(UpSetR)

```

## 0.2 Paths
```{r echo=F, message=FALSE, warning=FALSE}
fprefix <- "03_Deconvolution_CTVT_geneCorrelations"
wDir <- paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/")
dataDir <- paste0(wDir, "/031_Data/")
resDir <- paste0(wDir, "/033_Results/", fprefix, "/")
plotDir <- paste0(wDir, "/034_Plots/", fprefix, "/")
```

# 1 Data
## 1.2 Read in Adrian's purity table
Specificity = specific to a tumor (T) or host (H)
We have eT and eH. Of note, a transcript could be tumor specific but have expression only in the host because 

### 1.2.1 Data
```{r}
adrianDf <- read_csv(paste0(dataDir, "ctvt_rnaseq_expression-2.csv"))
head(adrianDf)

adrianCorr <- read.xlsx(paste0(dataDir, "CTVT_expression_purity_correlations.xlsx")) %>%
  dplyr::rename("Symbol"="gene")


### For each unique gene symbol, count number of times SNP is host or tumor specfic
intersectDf <- adrianDf  %>%
  group_by(GeneID, 
           Symbol) %>%
  summarise(
    Specificity_T_count = sum(eT > 0 & eH == 0),
    Specificity_H_count = sum(eH > 0 & eT == 0),
    .groups = "drop"
  )

### Binary H markers = 1,343 genes for which all SNPs are host specific. 
binaryHost <- intersectDf %>%
  filter(Specificity_T_count == 0 & Specificity_H_count > 0)

### Binary T markers = 4,639 genes for which all SNPs are tumor specific. 
binaryTumor <- intersectDf %>%
  filter(Specificity_H_count == 0 & Specificity_T_count > 0)

### How many genes are shared between both groups (i.e. some SNPs are host specific and some are tumor specific)
shared <- intersectDf %>%
  filter(Specificity_T_count > 0 & Specificity_H_count > 0 )
```

### 1.2.2 Data explanation
```{r}
# example <- adrianDf %>%
#   filter(Symbol == "CD19")
# example[1,]

# > table(example$Specificity)
#   H   T 
# 136 132 

# In 1208Ta, 18422072,	Ref - G,	Alt - A,	ENSCAFG00000017303, Specificity - H
# nAT = 2, nBT = 0. Both alleles are Ref - G in tumor. 
# nAH = 1, nBH = 1. 1x G and 1x A in host. Therefore, alt is specific to host. 
# eB = 10, therefore, in host, nA = 10 and nB = 10. 
# eB + eA = 16, but 16-20 doesn't work then?


# example[2,]
# 18425358	T	C	ENSCAFG00000017303, CD19, Spec H
# nAT = 2, nBT = 0. Both alleles are Ref - T in tumor. 
# nAH = 1, nBH = 1. 1x T and 1x C in host. Therefore, alt is specific to host. 
# eB = 5, therefore, in host, nA = 5 and nB = 5. 
# eB + eA = 26, so 26-10 = 16 expr in tumor * purity = 13.62... = 0.85 purity. and 10 * purity = 12.56...?

# example <- adrianDf %>%
#   filter(Symbol == "EPO")
# example[10,]
# head()
```

# 2 Tumor-specificity
## 2.3 eT > 0 and eH == 0 
### 2.3.1 Dataframes
#### 2.3.1.1 SNPs
! Of note, we need to group the table by location and Alt/Ref to count the number of samples with a unique SNP.
```{r}
# Summarize counts of Specificity = "T" and "H" per GeneID and Symbol
tumorSpecDf <- adrianDf %>%
  filter(eH == 0 & eT > 0) %>%
  group_by(GeneID, Symbol) %>%
  mutate("Annotation"="tumorOnly")

tumorSNPs <- unique(tumorSpecDf$Location)
tumorGenes <- unique(tumorSpecDf$Symbol)
tumorSymbols <- unique(tumorSpecDf$GeneID)
# table(tumorSpecDf$GeneID)

## There are 45 unique SNP's where there are more than 1 SNV per tumor. 
tumorSpecDfDuplicates <- tumorSpecDf %>%
  group_by(Tumour, Location) %>%
  filter(n() > 1) %>%
  ungroup()

# Example:
# head(tumorSpecDfDuplicates)
# 1 1208Ta 7:5641312   7      5641312 G     A 
# 2 1208Ta 7:5641312   7      5641312 G     T     


tumorSpecSampleCountDf <- tumorSpecDf %>%
  # distinct(Location, Tumour) %>%     # Keep one row per (Location, Tumour) pair
  group_by(Location, GeneID, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEt = mean(eT), 
            medianEt = median(eT))

# range(tumorSpecSampleCountDf$sampleCount)
# [1]  1 80

tumorSpecSampleAltCountDf <- tumorSpecDf %>%
  # distinct(Location, Tumour) %>%     # Keep one row per (Location, Tumour) pair
  group_by(Location, GeneID, Alt, Ref, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEt = mean(eT), 
            medianEt = median(eT))

# range(tumorSpecSampleAltCountDf$sampleCount)
# [1]  1 42
```

#### 2.3.1.2 Genes
Read in binary tumor df! (Section 4)
There are 4,639 unique genes, which can correspond to many SNPs. 
Merge data from adrianCorr to these tables. There is one purity corr value per gene, over all SNPs (because these are binary). 
Therefore, the alt/non alt doesn't work merging with purity, or we just repeat values. 
! Of note, a small set of genes (7) do not have an associated purity correlation. 
```{r}
tumorSpecSampleGeneDf <- tumorSpecDf %>%
  filter(GeneID %in% binaryTumor$GeneID) %>%  
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEt = mean(eT), 
            medianEt = median(eT)) %>%
  left_join(adrianCorr, 
            by = join_by(Symbol))

tumorSpecSampleAltGeneDf <- tumorSpecDf %>%
  filter(GeneID %in% binaryTumor$GeneID) %>%             #
  group_by(GeneID, Alt, Ref, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEt = mean(eT), 
            medianEt = median(eT)) %>%
  left_join(adrianCorr,
            by = join_by(Symbol))

# length(unique(tumorSpecSampleGeneDf$GeneID)) [1] 4639
```



#### 2.3.1.3 Subset SNPs 
Instead of filtering on SNPs first and checking what ones correspond to unique host or tumor genes --> group df by Symbol and select those where either all SNPs are Host or Tumor specific. 
```{r}
# Summarize counts of Specificity = "T" and "H" per GeneID and Symbol
tumorSpecDfStrict <- adrianDf %>%
  group_by(Symbol) %>%
  filter(all(eH == 0 & eT > 0)) %>%
  mutate("Annotation"="tumorOnlyStrict") %>%
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  summarise(sampleCount = n(),
            meanEt = mean(eT),
            medianEt = median(eT)) %>%
  left_join(adrianCorr,
            by = join_by(Symbol))
```

### 2.3.2 Visualizations
#### 2.3.2.1 SNPs
##### 2.3.2.1.1 Unique SNPs + Alt SNPs: histogram
```{r}
snpHisto <- ggplot(tumorSpecSampleCountDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueSNPs_sampleCount.pdf"))
print(snpHisto)
dev.off()

snpAltHisto <- ggplot(tumorSpecSampleAltCountDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueAltSNPs_sampleCount.pdf"))
print(snpAltHisto)
dev.off()
```
##### 2.3.2.1.2 Unique SNPs: mean eT vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecSampleCountDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             3000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             3000,
                             medianEt), 
         meanLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueSNPs_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

##### 2.3.2.1.3 Unique SNPs+Alt: mean eT vs # samples
! Of note, this does not mean that the tumor = alt !
```{r}
tempPlotDf <- tumorSpecSampleAltCountDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanAltLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA),
         medianAltLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit)) +
  geom_point() + 
  geom_label(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = .1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) + 
  theme_minimal()


medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueAltSNPs_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.2 Genes
##### 2.3.2.2.1 Unique Genes + Alt SNPs: histogram
```{r}
geneHisto <- ggplot(tumorSpecSampleGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=10, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueGenes_sampleCount.pdf"))
print(geneHisto)
dev.off()

geneAltHisto <- ggplot(tumorSpecSampleAltGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueAltGenes_sampleCount.pdf"))
print(geneAltHisto)
dev.off()
```
##### 2.3.2.2.2 Unique Genes: mean eT vs purity
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecSampleGeneDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 3000, meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 3000, medianEt), 
         meanLabel = ifelse((meanEt > 1000 & purity_correlation < -0.60) | 
                            (meanEt > 1000 & purity_correlation > 0.60), Symbol, NA),
         medianLabel = ifelse((medianEt > 1000 & purity_correlation < -0.60) | 
                              (medianEt > 1000 & purity_correlation > 0.60), Symbol, NA),
         ,
        colorGroup = case_when(
            purity_correlation > 0.5 ~ "high_pos",
            purity_correlation < -0.5 ~ "high_neg",
            TRUE ~ "neutral"
        ),
        fillColor = case_when(
            colorGroup == "high_pos" ~ meanEtEdit,   # will map to red
            colorGroup == "high_neg" ~ -meanEtEdit,  # will map to blue (note the minus for inversion)
            TRUE ~ NA_real_
        )
    )

tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, meanEtEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, meanEtEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

meanPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = meanEtEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Tumor-Specific Genes (n=4,639)",
    x = "Purity Correlation", 
    y = "Mean eT"
  )



tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, medianEtEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, medianEtEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = medianEtEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Tumor-Specific Genes (n=4,639)",
    x = "Purity Correlation", 
    y = "Median eT"
  )


pdf(paste0(plotDir, "/volcano/", fprefix, "_tumor_uniqueGenes_eT_vs_purity.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()

```

##### 2.3.2.2.3 Unique Genes: mean eT vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecSampleGeneDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA), 
         purityCol = ifelse(purity_correlation > 0, 
                            "Positive", 
                            "Negative"))

tempPlotDf$purityCol <- factor(tempPlotDf$purityCol, 
                               c("Positive", "Negative"))

(meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(0.15, "lines"),
    label.size = 0.05,
    color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  scale_color_paletteer_d("ggsci::nrc_npg") +
  theme_minimal() )

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(00, "lines"),
    label.size = 0.05,
     color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()+
  scale_color_paletteer_d("ggsci::nrc_npg") 

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueGenes_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```





##### 2.3.2.2.4 Subset genes
### 2.3.3 Visualizations - Strict
#### 2.3.2.1 Unique SNPs + Alt SNPs: histogram
```{r}
snpHisto <- ggplot(tumorSpecDfStrict, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_strict_uniqueSNPs_sampleCount.pdf"))
print(snpHisto)
dev.off()

snpAltHisto <- ggplot(tumorSpecDfStrict, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_strict_uniqueAltSNPs_sampleCount.pdf"))
print(snpAltHisto)
dev.off()
```
#### 2.3.2.2 Unique SNPs: mean eT vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecDfStrict %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_strict_uniqueSNPs_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.3 Unique SNPs+Alt: mean eT vs # samples
! Of note, this does not mean that the tumor = alt !
```{r}
tempPlotDf <- tumorSpecDfStrict %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanAltLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA),
         medianAltLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit)) +
  geom_point() + 
  geom_label(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = .1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) + 
  theme_minimal()


medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_strict_uniqueAltSNPs_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.1 Unique Genes + Alt SNPs: histogram
```{r}
geneHisto <- ggplot(tumorSpecSampleGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=10, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueGenes_sampleCount.pdf"))
print(geneHisto)
dev.off()

geneAltHisto <- ggplot(tumorSpecSampleAltGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_uniqueAltGenes_sampleCount.pdf"))
print(geneAltHisto)
dev.off()
```

#### 2.3.2.3 Unique Genes: mean eT vs purity
! Make this a volcano plot!
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecDfStrict %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 3000, meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 3000, medianEt), 
         meanLabel = ifelse((meanEt > 1000 & purity_correlation < -0.60) | 
                            (meanEt > 1000 & purity_correlation > 0.60), Symbol, NA),
         medianLabel = ifelse((medianEt > 1000 & purity_correlation < -0.60) | 
                              (medianEt > 1000 & purity_correlation > 0.60), Symbol, NA),
         ,
        colorGroup = case_when(
            purity_correlation > 0.5 ~ "high_pos",
            purity_correlation < -0.5 ~ "high_neg",
            TRUE ~ "neutral"
        ),
        fillColor = case_when(
            colorGroup == "high_pos" ~ meanEtEdit,   # will map to red
            colorGroup == "high_neg" ~ -meanEtEdit,  # will map to blue (note the minus for inversion)
            TRUE ~ NA_real_
        )
    )

tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, meanEtEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, meanEtEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

meanPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = meanEtEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Tumor-Specific Genes (n=58)",
       x = "Purity Correlation", 
       y = "Median eT")



tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, medianEtEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, medianEtEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = medianEtEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Tumor-Specific Genes (n=58)",
       x = "Purity Correlation", 
       y = "Median eT")



pdf(paste0(plotDir, "/volcano/", fprefix, "_tumor_strict_uniqueGenes_eT_vs_purity.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.2 Unique Genes: mean eT vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- tumorSpecSampleGeneDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            Symbol, 
                            NA), 
         purityCol = ifelse(purity_correlation > 0, 
                            "Positive", 
                            "Negative"))

tempPlotDf$purityCol <- factor(tempPlotDf$purityCol, 
                               c("Positive", "Negative"))

(meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(0.15, "lines"),
    label.size = 0.05,
    color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  scale_color_paletteer_d("ggsci::nrc_npg") +
  theme_minimal() )

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(00, "lines"),
    label.size = 0.05,
     color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()+
  scale_color_paletteer_d("ggsci::nrc_npg") 

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueGenes_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```
#### 2.3.2.4 Unique SNPs+Alt: mean eT vs # samples
! Of note, this does not mean that the tumor = alt !
```{r}
tempPlotDf <- tumorSpecSampleAltCountDf %>%
  mutate(meanEtEdit = ifelse(meanEt == Inf, 
                             7000,
                             meanEt),
         medianEtEdit = ifelse(medianEt == Inf, 
                             7000,
                             medianEt), 
         meanAltLabel = ifelse(meanEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA),
         medianAltLabel = ifelse(medianEt > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEtEdit)) +
  geom_point() + 
  geom_label(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = .1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) + 
  theme_minimal()


medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEtEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueAltSNPs_eT_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```





# 3 Host
## 3.1 eH > 0 and eT == 0
### 2.3.1 Dataframes
#### 2.3.1.1 SNPs
! Of note, we need to group the table by location and Alt/Ref to count the number of samples with a unique SNP.
```{r}
hostSpecDf <- adrianDf %>%
  filter(eH > 0 & eT == 0) %>%
  group_by(GeneID, Symbol) %>%
  mutate("Annotation"="hostOnly")

hostSNPs <- unique(hostSpecDf$Location)
hostGenes <- unique(hostSpecDf$Symbol)
hostSymbols <- unique(hostSpecDf$GeneID)

## There are 54 unique SNP's where there are more than 1 SNV per host. 
hostSpecDfDuplicates <- hostSpecDf %>%
  group_by(Tumour, Location) %>%
  filter(n() > 1) %>%
  ungroup()

hostSpecSampleCountDf <- hostSpecDf %>%
  
  group_by(Location, GeneID, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEh = mean(eH), 
            medianEh = median(eH))

# range(hostSpecSampleCountDf$sampleCount)
# [1]  1 69

hostSpecSampleAltCountDf <- hostSpecDf %>%
  
  group_by(Location, GeneID, Alt, Ref, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEh = mean(eH), 
            medianEh = median(eH))

# range(hostSpecSampleAltCountDf$sampleCount)
# [1]  1 42
```

#### 2.3.1.2 Genes
Read in binary host df! (Section 4)
There are 4,639 unique genes, which can correspond to many SNPs. 
Merge data from adrianCorr to these tables. There is one purity corr value per gene, over all SNPs (because these are binary). 
Therefore, the alt/non alt doesn't work merging with purity, or we just repeat values. 
! Of note, a small set of genes (7) do not have an associated purity correlation. 
```{r}
hostSpecSampleGeneDf <- hostSpecDf %>%
  filter(GeneID %in% binaryHost$GeneID) %>%  
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEh = mean(eH), 
            medianEh = median(eH)) %>%
  left_join(adrianCorr, 
            by = join_by(Symbol))

hostSpecSampleAltGeneDf <- hostSpecDf %>%
  filter(GeneID %in% binaryHost$GeneID) %>%             #
  group_by(GeneID, Alt, Ref, Symbol) %>%             # Group by SNP (Location)
  summarise(sampleCount = n(), 
            meanEh = mean(eH), 
            medianEh = median(eH)) %>%
  left_join(adrianCorr,
            by = join_by(Symbol))

# length(unique(hostSpecSampleGeneDf$GeneID)) [1] 4639
```

#### 2.3.1.3 Subset SNPs 
Instead of filtering on SNPs first and checking what ones correspond to unique host or tumor genes --> group df by Symbol and select those where either all SNPs are Host or Tumor specific. 
```{r}
hostSpecDfStrict <- adrianDf %>%
  group_by(Symbol) %>%
  filter(all(eH > 0 & eT == 0)) %>%
  mutate("Annotation"="humanOnlyStrict") %>%
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  summarise(sampleCount = n(),
            meanEh = mean(eH),
            medianEh = median(eH)) %>%
  left_join(adrianCorr,
            by = join_by(Symbol))

### Do any genes overlap? No!
intersect(tumorSpecDfStrict$Symbol, hostSpecDfStrict$Symbol)
```

### 2.3.2 Visualizations
#### 2.3.2.1 Unique SNPs + Alt SNPs: histogram
```{r}
snpHisto <- ggplot(hostSpecSampleCountDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_host_uniqueSNPs_sampleCount.pdf"))
print(snpHisto)
dev.off()

snpAltHisto <- ggplot(hostSpecSampleAltCountDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_host_uniqueAltSNPs_sampleCount.pdf"))
print(snpAltHisto)
dev.off()
```
#### 2.3.2.2 Unique SNPs: mean eH vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- hostSpecSampleCountDf %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 
                             7000,
                             meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 
                             7000,
                             medianEh), 
         meanLabel = ifelse(meanEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_host_uniqueSNPs_eH_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.3 Unique SNPs+Alt: mean eH vs # samples
! Of note, this does not mean that the host = alt !
```{r}
tempPlotDf <- hostSpecSampleAltCountDf %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 
                             7000,
                             meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 
                             7000,
                             medianEh), 
         meanAltLabel = ifelse(meanEh > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA),
         medianAltLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit)) +
  geom_point() + 
  geom_label(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = .1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) + 
  theme_minimal()


medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_host_uniqueAltSNPs_eH_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.1 Unique Genes + Alt SNPs: histogram
```{r}
geneHisto <- ggplot(hostSpecSampleGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=10, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()

pdf(paste0(plotDir, "/histogram/", fprefix, "_host_uniqueGenes_sampleCount.pdf"))
print(geneHisto)
dev.off()

geneAltHisto <- ggplot(hostSpecSampleAltGeneDf, aes(x=sampleCount)) +
  geom_histogram(binwidth=1, fill="#69b3a2", color="#e9ecef", alpha=0.9) +
  theme_minimal()


pdf(paste0(plotDir, "/histogram/", fprefix, "_host_uniqueAltGenes_sampleCount.pdf"))
print(geneAltHisto)
dev.off()
```
#### 2.3.2.3 Unique Genes: mean eH vs purity
! Make this a volcano plot!
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- hostSpecSampleGeneDf %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 7000, meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 7000, medianEh), 
         meanLabel = ifelse((meanEh > 1000 & purity_correlation < -0.60) | 
                            (meanEh > 1000 & purity_correlation > 0.60), Symbol, NA),
         medianLabel = ifelse((medianEh > 1000 & purity_correlation < -0.60) | 
                              (medianEh > 1000 & purity_correlation > 0.60), Symbol, NA),
         ,
        colorGroup = case_when(
            purity_correlation > 0.5 ~ "high_pos",
            purity_correlation < -0.5 ~ "high_neg",
            TRUE ~ "neutral"
        ),
        fillColor = case_when(
            colorGroup == "high_pos" ~ meanEhEdit,   # will map to red
            colorGroup == "high_neg" ~ -meanEhEdit,  # will map to blue (note the minus for inversion)
            TRUE ~ NA_real_
        )
    )

tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, meanEhEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, meanEhEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

meanPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = meanEhEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(
    title = "Unique Host-Specific Genes (n=1,343)",
    x = "Purity Correlation", 
    y = "Mean eH"
  )



tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, medianEhEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, medianEhEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = medianEhEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(
    title = "Unique Host-Specific Genes (n=1,343)",
    x = "Purity Correlation", 
    y = "Median eH"
  )


pdf(paste0(plotDir, "/volcano/", fprefix, "_host_uniqueGenes_eH_vs_purity.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```

#### 2.3.2.2 Unique Genes: mean eH vs # samples
! Of note, making Inf arbitrarily large (7,000)
```{r}
tempPlotDf <- hostSpecSampleGeneDf %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 
                             7000,
                             meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 
                             7000,
                             medianEh), 
         meanLabel = ifelse(meanEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA),
         medianLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA), 
         purityCol = ifelse(purity_correlation > 0, 
                            "Positive", 
                            "Negative"))

tempPlotDf$purityCol <- factor(tempPlotDf$purityCol, 
                               c("Positive", "Negative"))

(meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(0.15, "lines"),
    label.size = 0.05,
    color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  scale_color_paletteer_d("ggsci::nrc_npg") +
  theme_minimal() )

medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEhEdit, color = purityCol, size=abs(purity_correlation))) +
  geom_point() + 
  geom_label_repel(
    aes(label = medianLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    # label.padding = unit(00, "lines"),
    label.size = 0.05,
     color = "white",
    fill = "black",
    min.segment.length = 0,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()+
  scale_color_paletteer_d("ggsci::nrc_npg") 

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueGenes_eH_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```
#### 2.3.2.4 Unique SNPs+Alt: mean eH vs # samples
! Of note, this does not mean that the host = alt !
```{r}
tempPlotDf <- hostSpecSampleAltCountDf %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 
                             7000,
                             meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 
                             7000,
                             medianEh), 
         meanAltLabel = ifelse(meanEh > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA),
         medianAltLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            paste0(Symbol, "_", Ref, "->", Alt), 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit)) +
  geom_point() + 
  geom_label(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = .1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) + 
  theme_minimal()


medianPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=medianEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanAltLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

pdf(paste0(plotDir, "/scatter/", fprefix, "_uniqueAltSNPs_eH_vs_sampleCount.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```
### 2.3.3 Visualizations - Strict
#### 2.3.3.3 Unique Genes: mean eT vs purity
```{r}
tempPlotDf <- hostSpecDfStrict %>%
  mutate(meanEhEdit = ifelse(meanEh == Inf, 3000, meanEh),
         medianEhEdit = ifelse(medianEh == Inf, 3000, medianEh), 
         meanLabel = ifelse((meanEh > 1000 & purity_correlation < -0.60) | 
                            (meanEh > 1000 & purity_correlation > 0.60), Symbol, NA),
         medianLabel = ifelse((medianEh > 1000 & purity_correlation < -0.60) | 
                              (medianEh > 1000 & purity_correlation > 0.60), Symbol, NA),
         ,
        colorGroup = case_when(
            purity_correlation > 0.5 ~ "high_pos",
            purity_correlation < -0.5 ~ "high_neg",
            TRUE ~ "neutral"
        ),
        fillColor = case_when(
            colorGroup == "high_pos" ~ meanEhEdit,   # will map to red
            colorGroup == "high_neg" ~ -meanEhEdit,  # will map to blue (note the minus for inversion)
            TRUE ~ NA_real_
        )
    )

tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, meanEhEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, meanEhEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

meanPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = meanEhEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Host-Specific Genes (n=52)",
       x = "Purity Correlation", 
       y = "Median eH")



tempPlotDf <- tempPlotDf %>%
  mutate(
    fillPos = ifelse(purity_correlation > 0.5, medianEhEdit, NA),
    fillNeg = ifelse(purity_correlation < -0.5, medianEhEdit, NA)
  )

# Max for scaling
max_pos <- max(tempPlotDf$fillPos, na.rm = TRUE)
max_neg <- max(tempPlotDf$fillNeg, na.rm = TRUE)

medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = medianEhEdit)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # NEGATIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation < -0.5),
             aes(fill = fillNeg, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "- Corr.",
    low = "lightblue", high = "blue4",
    limits = c(0, max_neg), na.value = "grey90"
  ) +

  # POSITIVE correlation points layer
  new_scale_fill() +
  geom_point(data = subset(tempPlotDf, purity_correlation > 0.5),
             aes(fill = fillPos, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "+ Corr.",
    low = "mistyrose", high = "darkred",
    limits = c(0, max_pos), na.value = "grey90"
  ) +

  theme_minimal() +
  labs(title = "Unique Host-Specific Genes (n=52)",
       x = "Purity Correlation", 
       y = "Median eH")



pdf(paste0(plotDir, "/volcano/", fprefix, "_host_strict_uniqueGenes_eH_vs_purity.pdf"), 
    height=6, 
    width=8)
print(meanPlot)
print(medianPlot)
dev.off()
```




# 4 Intersect
9,062 gene symbols have both host- and tumor-informative SNPs. 
We need to look at how many times they are host vs tumor informative. 
```{r}
### For each unique gene symbol, count number of times SNP is host or tumor specfic
intersectDf <- adrianDf  %>%
  group_by(GeneID, 
           Symbol) %>%
  summarise(
    Specificity_T_count = sum(eT > 0 & eH == 0),
    Specificity_H_count = sum(eH > 0 & eT == 0),
    .groups = "drop"
  )

### Binary H markers = 1,343 genes for which all SNPs are host specific. 
binaryHost <- intersectDf %>%
  filter(Specificity_T_count == 0 & Specificity_H_count > 0)

### Binary T markers = 4,639 genes for which all SNPs are tumor specific. 
binaryTumor <- intersectDf %>%
  filter(Specificity_H_count == 0 & Specificity_T_count > 0)

### How many genes are shared between both groups (i.e. some SNPs are host specific and some are tumor specific)
shared <- intersectDf %>%
  filter(Specificity_T_count > 0 & Specificity_H_count > 0 )
```




# 5 Merge with external data
### 1.3.1 HPA: Monaco
Trying to subset Monaco
Testing with min nTPM 20 = 7,765 genes. 
```{r}
monaco <- read_tsv("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/031_Data/HPA/rna_immune_cell_monaco.tsv") %>%
  dplyr::rename("Gene name"="geneName") %>%
  filter(geneName %in% binaryHost$Symbol & nTPM > 20)
```
# 6 Find All Markers
## 6.1 Convert SCE to Seurat
Only keeping non-bad clusters because I want "true" marker genes. 
```{r}
obj <- readRDS(paste0(dataDir, "SCE_concat_CTVT_includingAllGenes_logCorrecDone_vstCorrecDone_ClustDone2_ManAnot.rds"))

assay(obj, "counts") <- as.matrix(assay(obj, "counts"))
assay(obj, "logcounts") <- as.matrix(assay(obj, "logcounts"))
seuratObj <- as.Seurat(obj, counts = "counts", data = "logcounts")
seuratObj
Idents(seuratObj) <- "manual.log.anot.fine"
seuratObjNoBad <- subset(seuratObj, idents = "bad", invert=TRUE)
table(Idents(seuratObjNoBad))
DimPlot(seuratObj, reduction = "TSNE_log_corrected")
DimPlot(seuratObjNoBad, reduction = "TSNE_log_corrected")

rm(obj, seuratObj)
gc()


FeaturePlot(seuratObjNoBad, "ANXA1", reduction="TSNE_log_corrected")
FeaturePlot(seuratObjNoBad, "CXCL10", reduction="TSNE_log_corrected")
```

## 6.2 Find all markers
### 6.2.1 Wilcox
```{r}
# famWilcox <- FindAllMarkers(seuratObjNoBad, 
#                       group.by = "manual.log.anot.fine", 
#                       min.pct = 0.05)
# 
# write.xlsx(famWilcox, 
#            paste0(resDir, "findAllMarkers/", fprefix, "_seuratObjNoBad_famWilcox.xlsx"))
# 
# gc()
```

### 6.2.2 MAST
```{r}
# famMast <- FindAllMarkers(seuratObjNoBad, 
#                           test.use = "MAST",
#                           group.by = "manual.log.anot.fine", 
#                           min.pct = 0.05)
# write.xlsx(famMast, 
#            paste0(resDir, "findAllMarkers/", fprefix, "_seuratObjNoBad_famMast.xlsx"))
```

## 6.3 Tumor specific genes
### 6.3.1 Filter genes 
Testing positive genes from FAM with high purity correlations. 
```{r}
famWilcox <- read.xlsx(paste0(resDir, "findAllMarkers/", fprefix, "_seuratObjNoBad_famWilcox.xlsx"))
famWilcoxDegs <- famWilcox %>%
  filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster == "CTVT") 

intersecting <- intersect(famWilcoxDegs$gene, tumorSpecSampleGeneDf$Symbol) ### 33 genes intersecting
# intersectingStrict <- intersect(famWilcoxDegs$gene, tumorSpecDfStrict$Symbol) ### None overlap. 
```

### 6.3.2 Merge with purity correlations and eT
```{r}
mergedFamWilcoxDegsCTVT <- famWilcox %>%
  filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster=="CTVT") %>%
  filter(gene %in% tumorSpecSampleGeneDf$Symbol) %>%
  dplyr::dplyr::rename("Symbol"="gene") %>%
  left_join(tumorSpecSampleGeneDf) %>%
  mutate("Annotation"=paste0("tumorSpec_", cluster, "Deg")) 


# write.xlsx(mergedFamWilcoxDegsCTVT, 
#            paste0(resDir, fprefix, "_mergedFamWilcoxDegs_adrianTumorGenes.xlsx"))

ctvtSet <- mergedFamWilcoxDegsCTVT$Symbol
```
### 6.3.3 Feature plots
Merged and split by sample
```{r}
# for(g in intersecting){
# 
#   feat <- FeaturePlot(seuratObjNoBad,
#                       reduction = "TSNE_log_corrected",
#                       features=g) +
# 
#     ggtitle(paste0("tumorSpecDf + Wilcox DEG - ", g))
#   
#   featSamp <- FeaturePlot(seuratObjNoBad, 
#                           reduction = "TSNE_log_corrected",
#                           features=g, 
#                           split.by = "Sample") 
#   
#   pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_tumorSpecDf_famWilcoxDegs_", g,"_expression.pdf"),
#       height=4,
#       width=4)
# 
#   print(feat)
# 
#   dev.off()
#   
#   
#   
#   pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix,  "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_tumorSpecDf_famWilcoxDegs_", g,"_expressionBySample.pdf"), 
#       height=4, 
#       width=14)
#   
#   print(featSamp) 
#   
#   dev.off()
#   
# }
```


## 6.4 Host specific genes
### 6.4.1 Feature plots
```{r}
# ###### 2.1.4 Loop through NON CTVT FAM DEG's --> check against genetic deconv list. Test positive genes first. 
# for(ctype in unique(famWilcox$cluster)){
#   if(ctype=="CTVT"){
#     next
#   }
#   
#   tempFamWilcoxDegs <- famWilcox %>%
#     filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster == ctype) 
#   
#   tempIntersecting <- intersect(tempFamWilcoxDegs$gene, hostSpecSampleGeneDf$Symbol)
#   
#   print(paste0("Cell type: ", ctype, " - Intersecting: ", length(tempIntersecting)))
#   
#   tempPlotDir <- paste0("hostSpecificIntersect", ctype)
#   
#   for(tg in tempIntersecting){
#     
#     feat <- FeaturePlot(seuratObjNoBad,
#                         reduction = "TSNE_log_corrected",
#                         features=tg) +
#       
#       ggtitle(paste0("hostSpecDf + ", ctype, " Wilcox DEG - ", tg))
#     
#     featSamp <- FeaturePlot(seuratObjNoBad, 
#                             reduction = "TSNE_log_corrected",
#                             features=tg, 
#                             split.by = "Sample") 
#     
#     pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_tumorSpecDf_famWilcoxDegs_", tg,"_expression.pdf"),
#         height=4,
#         width=4)
#     
#     print(feat)
#     
#     dev.off()
#     
#     
#     
#     pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_tumorSpecDf_famWilcoxDegs_", tg,"_expressionBySample.pdf"), 
#         height=4, 
#         width=14)
#     
#     print(featSamp) 
#     
#     dev.off()
#     
#   }
#   
# }
```

### 6.4.2 Merge with purity correlations
```{r}
mergedFamWilcoxDegsAdrian <- data.frame()


for(ctype in unique(famWilcox$cluster)){
  if(ctype=="CTVT"){
    next
  }
  
  temp <- famWilcox %>%
    filter(p_val_adj < 0.01 & avg_log2FC > 2.5& cluster==ctype) %>%
    filter(gene %in% hostSpecSampleGeneDf$Symbol) %>%
    dplyr::dplyr::rename("Symbol"="gene") %>%
    left_join(hostSpecSampleGeneDf) %>%
    mutate("Annotation"=paste0("hostSpec_", ctype, "Deg"))
  
  mergedFamWilcoxDegsAdrian <- rbind(mergedFamWilcoxDegsAdrian, temp)
  
}
```


## 6.5 Candidate gene lists
### 6.5.1 Gene sets
```{r}
mergedFamWilcoxDegsCTVT <- famWilcox %>%
  filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster=="CTVT") %>%
  filter(gene %in% tumorSpecSampleGeneDf$Symbol) %>%
  dplyr::rename("Symbol"="gene") %>%
  left_join(tumorSpecSampleGeneDf) %>%
  mutate("Annotation"=paste0("tumorSpec_", cluster, "Deg")) 

# 17 genes are negative correlated with tumor purity and a Bcell DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Bcells" & purity_correlation < -0.5)))
candBcell <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Bcells" & purity_correlation < -0.5) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

bcellSet <- candBcell$Symbol

# 36 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Myeloid" & purity_correlation < -0.5)))
candMyeloid <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Myeloid" & purity_correlation < -0.5) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

myeloidSet <- candMyeloid$Symbol

# 32 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "CAF.Endo" & purity_correlation < -0.5)))
candCAFEndo <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "CAF.Endo" & purity_correlation < -0.5) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

cafendoSet <- candCAFEndo$Symbol


# 21 genes are negatively correlated with tumor purity and a Plasma DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Plasma" & purity_correlation < -0.5)))
candPlasma <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Plasma" & purity_correlation < -0.5) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

plasmaSet <- candPlasma$Symbol

# 9 genes are negatively correlated with tumor purity and a Tcells DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Tcells" & purity_correlation < -0.5)))
candTcells <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Tcells" & purity_correlation < -0.5) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

tcellSet <- candTcells$Symbol
```

### 6.5.2 Upset
```{r}
geneSets <- list("ctvt"=ctvtSet,
             "bcell"=bcellSet,
             "myeloid"=myeloidSet,
             "caf.endo"=cafendoSet,
             "plasma"=plasmaSet,
             "tcell"=tcellSet)


genes <- sort(unique(unlist(geneSets)))

sets <- t(+sapply(geneSets, "%in%", x = genes)) %>%
  t() %>%
  as.data.frame()

rownames(sets) <- genes

sets <- sets %>%
  rownames_to_column("Gene")

upsetPlot <- upset(sets,
      sets = c("ctvt", "bcell", "myeloid", "caf.endo", "plasma", "tcell"),
      # keep.order = TRUE,
      nintersects = NA, # show all intersections
      order.by = "freq") # optional, could be degree or freq

upsetPlot <- upset(
  sets,
  sets = c("ctvt", "caf.endo", "myeloid", "plasma", "tcell", "bcell"),
  keep.order = TRUE,
  nintersects = NA,
  sets.bar.color = c(
    ctvt = "#E41A1C",
    caf.endo = "gold",
    myeloid = "#4DAF4A",
    plasma = "skyblue",
    tcell = "cornflowerblue",
    bcell = "pink"
  )
)


uniqueGenes <- sets %>%
  column_to_rownames("Gene") %>%
  mutate(rowsum = rowSums(.)) %>%
  filter(rowsum==1)


upsetData <- sets %>%
  unite(col = "intersection", -c("Gene"), sep = "") %>%
  group_by(intersection) %>%
  summarise(list = list(Gene)) %>%
  mutate(list = setNames(list, intersection)) %>%
  pull(list)

write.xlsx(sets, paste0(resDir, fprefix, "_geneCorrelations_famDegs.xlsx"))

pdf(paste0(plotDir, "/upset/", fprefix, "_hostTumorGenes_famDegs_intersection.pdf"), 
    height=6, 
    width=8)
print(upsetPlot)
dev.off()
```

### 6.5.3 Unique gene sets
```{r}
uniqueGenesFinal <- uniqueGenes %>%
  dplyr::select(-rowsum)

# Initialize a list to store unique genes per cell type
uniqueGeneLists <- list()

# Loop through each column (cell type)
for (cell_type in colnames(uniqueGenesFinal)) {
  # Identify genes that are only expressed (1) in this cell type
  unique_genes <- rownames(uniqueGenesFinal)[
    uniqueGenesFinal[, cell_type] == 1 & rowSums(uniqueGenesFinal) == 1
  ]
  
  # Store in the list
  uniqueGeneLists[[cell_type]] <- unique_genes
}
```



# 7 Median eT vs eH
## 7.1 Data
New approach for marker gene selection: 
- Group by gene, take median eT and eH --> probably more complete with purity corr. by gene as well. 
- Of note, max eT = Inf --> replacing with 60000. 
```{r}
# snpDf <- 
geneDf <- adrianDf %>%
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  mutate(etEdit = ifelse(eT == Inf, 
                             50000,
                             eT)) %>%
  summarise(sampleCount = n(), 
            medianEh = median(eH), 
            medianEt = median(eT),
            medianEtEdit = median(etEdit)) %>%
  left_join(adrianCorr, 
            by = join_by(Symbol))
```

## 7.2 Plot
### 7.2.1 Median eT vs eH 
```{r}
plotDf <- geneDf %>%
  mutate(medianEhLog = log1p(medianEh), 
         medianEtLog = log1p(medianEt),
         medianEtEditLog = log10(medianEtEdit),
         labelTH = ifelse(medianEh > 1000 & medianEt > 1000,
                            Symbol, 
                            NA),
         labelNeither = ifelse(medianEh < 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelH =  ifelse(medianEh > 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelT = ifelse(medianEh < 1000 & medianEt > 1000,
                            Symbol, 
                            NA))

medianPlot <- ggplot(plotDf, aes(x = medianEhLog, y = medianEtLog)) +
  geom_point() +
  theme_minimal() +
  labs(title = "eT vs eH",
    x = "log1p(median(eH))", 
    y = "logq1p(median(eT))"
  ) +
  geom_label_repel(
    aes(label = labelTH),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  # geom_label_repel(
  #   aes(label = labelNeither),
  #   nudge_x = 0.1, 
  #   nudge_y = 0.1,
  #   label.padding = unit(0.15, "lines"),
  #   label.size = 0.1,
  #   color = "black",
  #   fill = "firebrick2",
  #   min.segment.length = .5,
  #   segment.color = "firebrick2"  # line color from point to label
  # ) +
  
  geom_label_repel(
    aes(label = labelT),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "skyblue",
    min.segment.length = .5,
    segment.color = "skyblue"  # line color from point to label
  ) +
  geom_label_repel(
    aes(label = labelH),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "gold",
    min.segment.length = .5,
    segment.color = "gold"  # line color from point to label
  )

pdf(paste0(plotDir, "/scatter/", fprefix, "_medianEtEh_labelled.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```

#### 7.2.1.1 Logscale
```{r}
medianPlot <- ggplot(plotDf, aes(x = medianEh, y = medianEt)) +
  geom_point() +
  scale_x_log10(guide = "axis_logticks", limits=c(10^-3, 10^4)) +
  scale_y_log10(guide = "axis_logticks", limits=c(10^-3, 10^4)) + 
  theme_minimal() +
  labs(title = "eT vs eH",
    x = "median(eH)", 
    y = "median(eT)"
  )

medianLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

```
#### 7.2.1.2 Absolute

### 7.2.2 Distribution of delta eT and eH vs purity correlation
#### 7.2.2.1 Delta eT and eH
```{r}
plotDf <- geneDf %>%
  mutate(deltaEtEh = medianEt-medianEh,
         medianEhLog = log1p(medianEh), 
         medianEtLog = log1p(medianEt),
         medianEtEditLog = log10(medianEtEdit),
         labelTH = ifelse(medianEh > 1000 & medianEt > 1000,
                            Symbol, 
                            NA),
         labelNeither = ifelse(medianEh < 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelH =  ifelse(medianEh > 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelT = ifelse(medianEh < 1000 & medianEt > 1000,
                            Symbol, 
                            NA))

write.xlsx(plotDf, paste0(resDir, fprefix, "_deltaEtEh.xlsx"))
```
#### 7.2.2.2 Purity vs delta
```{r}
tempPlotDf <- plotDf %>%
  rename("sampleCount"="variantCount") %>%
  mutate(
    fillPos = medianEtEdit,
    fillNeg = medianEh
  )

# Max for scaling
q1 <- subset(tempPlotDf, purity_correlation > 0 & deltaEtEh > 0)
maxq1 <- max(q1$deltaEtEh)
q1Label <- 

q2 <- subset(tempPlotDf, purity_correlation > 0 & deltaEtEh < 0)
maxq2 <- min(q2$deltaEtEh)

q3 <- subset(tempPlotDf, purity_correlation < 0 & deltaEtEh > 0)
maxq3 <- min(q3$deltaEtEh)

q4 <- subset(tempPlotDf, purity_correlation < 0 & deltaEtEh < 0)
maxq4 <- min(q4$deltaEtEh)


medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = deltaEtEh)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # POS POS correlation points layer
  new_scale_fill() +
  geom_point(data = q1,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "mistyrose", high = "maroon",
    limits = c(0, 5000), na.value = "grey90"
  ) +
   # NEG POS correlation points layer
    new_scale_fill() +
  geom_point(data = q2,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    high = "#ffd580", low = "firebrick2",
    limits = c(maxq2, 0), na.value = "grey90"
  ) +
  
  # NEG POS correlation points layer
    new_scale_fill() +
  geom_point(data = q3,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "#7ED6B7", high = "#1831F5",
    limits = c(0,  5000), na.value = "grey90"
  ) +
    # NEG POS correlation points layer
  new_scale_fill() +
  geom_point(data = q4,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "#7E53E6", high = "#D8BFD8",
    limits = c(maxq4, 0), na.value = "grey90"
  ) +
  theme_minimal() +
  labs(
    title = "",
    x = "Purity Correlation", 
    y = "Median eT - Median eH"
  ) 

pdf(paste0(plotDir, "/volcano/", fprefix, "_deltaEtEh_purity.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```
#### 7.2.2.3 Labeled purity vs delta
```{r}
# Filter points with abs(deltaEtEh) > 500 in Q1 and Q3
q1Label <- subset(q1, abs(deltaEtEh) > 200)
q4Label <- subset(q4, abs(deltaEtEh) > 200)

# Plot with labels
medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = deltaEtEh)) +
  # NEUTRAL points
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = variantCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # Q1
  new_scale_fill() +
  geom_point(data = q1,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "mistyrose", high = "maroon", limits = c(0, 5000), na.value = "grey90") +

  # Q2
  new_scale_fill() +
  geom_point(data = q2,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(high = "#ffd580", low = "firebrick2", limits = c(maxq2, 0), na.value = "grey90") +

  # Q3
  new_scale_fill() +
  geom_point(data = q3,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "#7ED6B7", high = "#1831F5", limits = c(0, 5000), na.value = "grey90") +

  # Q4
  new_scale_fill() +
  geom_point(data = q4,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "#7E53E6", high = "#D8BFD8", limits = c(maxq4, 0), na.value = "grey90") +

  # Labels for Q1 and Q4
  geom_text(data = q1Label, aes(label = Symbol), size = 3, vjust = -1, check_overlap = TRUE) +
  geom_text(data = q4Label, aes(label = Symbol), size = 3, vjust = -1, check_overlap = TRUE) +
  scale_size(range = c(1,10)) +

  theme_minimal() +
  labs(x = "Purity Correlation", y = "Median eT - Median eH")

pdf(paste0(plotDir, "/volcano/", fprefix, "_deltaEtEh_purity_labelled200.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```

## 7.3 With marker genes
### 7.3.1 Candidate gene lists
```{r}
# Filter points with abs(deltaEtEh) > 500 in Q1 and Q3

# q1 = tumor
q1Label100 <- subset(q1, abs(deltaEtEh) > 100)

# q4 = host
q4Label100 <- subset(q4, abs(deltaEtEh) > 100)
```

### 7.3.2 CTVT
#### 7.3.2.1 Filter genes 
Testing positive genes from FAM with high purity correlations. 
```{r}
famWilcox <- read.xlsx(paste0(resDir, "findAllMarkers/", fprefix, "_seuratObjNoBad_famWilcox.xlsx"))

famWilcoxDegs <- famWilcox %>%v
  filter(p_val_adj < 0.05 & avg_log2FC > 2.5 & cluster == "CTVT") 

intersecting <- intersect(famWilcoxDegs$gene, q1Label100$Symbol) ### 93 genes intersecting

# intersectingStrict <- intersect(famWilcoxDegs$gene, tumorSpecDfStrict$Symbol) ### None overlap. 
```

#### 7.3.2.2 Merge with purity correlations and eT
```{r}
mergedFamWilcoxDegsCTVT <- famWilcox %>%
  filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster=="CTVT") %>%
  filter(gene %in% q1Label100$Symbol) %>%
  dplyr::rename("Symbol"="gene") %>%
  left_join(q1Label100) %>%
  mutate("Annotation"=paste0("q1Label100_CTVTDeg")) 

write.xlsx(mergedFamWilcoxDegsCTVT,
           paste0(resDir, fprefix, "_mergedFamWilcoxDegs_q1100TumorGenes.xlsx"))

ctvtSet <- mergedFamWilcoxDegsCTVT$Symbol
```
#### 7.3.2.3 Feature plots
Merged and split by sample
```{r}
for(g in intersecting){

  feat <- FeaturePlot(seuratObjNoBad,
                      reduction = "TSNE_log_corrected",
                      features=g) +

    ggtitle(paste0("q1 + Wilcox DEG - ", g))

  featSamp <- FeaturePlot(seuratObjNoBad,
                          reduction = "TSNE_log_corrected",
                          features=g,
                          split.by = "Sample")

  pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_q1label100_famWilcoxDegs_", g,"_expression.pdf"),
      height=4,
      width=4)

  print(feat)

  dev.off()



  pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix,  "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_q1label100_famWilcoxDegs_", g,"_expressionBySample.pdf"),
      height=4,
      width=14)

  print(featSamp)

  dev.off()

}
```

### 7.3.2 Host specific genes
#### 7.3.2.1 Feature plots
```{r}
###### 2.1.4 Loop through NON CTVT FAM DEG's --> check against genetic deconv list. Test positive genes first.
for(ctype in unique(famWilcox$cluster)){
  if(ctype=="CTVT"){
    next
  }

  tempFamWilcoxDegs <- famWilcox %>%
    filter(p_val_adj < 0.05 & avg_log2FC > 2.5 & cluster == ctype)

  tempIntersecting <- intersect(tempFamWilcoxDegs$gene, q4Label100$Symbol)

  print(paste0("Cell type: ", ctype, " - Intersecting: ", length(tempIntersecting)))

  tempPlotDir <- paste0("hostSpecificIntersect", ctype)

  for(tg in tempIntersecting){

    feat <- FeaturePlot(seuratObjNoBad,
                        reduction = "TSNE_log_corrected",
                        features=tg) +

      ggtitle(paste0("q4Label100 + ", ctype, " Wilcox DEG - ", tg))

    featSamp <- FeaturePlot(seuratObjNoBad,
                            reduction = "TSNE_log_corrected",
                            features=tg,
                            split.by = "Sample")

    pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_q4Label100_famWilcoxDegs_", tg,"_expression.pdf"),
        height=4,
        width=4)

    print(feat)

    dev.off()



    pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_q4Label100_famWilcoxDegs_", tg,"_expressionBySample.pdf"),
        height=4,
        width=14)

    print(featSamp)

    dev.off()

  }

}
```

### 6.4.2 Merge with purity correlations
```{r}
mergedFamWilcoxDegsAdrian <- data.frame()


for(ctype in unique(famWilcox$cluster)){
  if(ctype=="CTVT"){
    next
  }
  
  temp <- famWilcox %>%
    filter(p_val_adj < 0.05 & avg_log2FC > 2.5& cluster==ctype) %>%
    filter(gene %in% q4Label100$Symbol) %>%
    dplyr::rename("Symbol"="gene") %>%
    left_join(q4Label100) %>%
    mutate("Annotation"=paste0("q4Label100_", ctype, "Deg"))
  
  mergedFamWilcoxDegsAdrian <- rbind(mergedFamWilcoxDegsAdrian, temp)
  
}

allDegs <- rbind(mergedFamWilcoxDegsAdrian, mergedFamWilcoxDegsCTVT)

write.xlsx(allDegs,
           paste0(resDir, fprefix, "_mergedFamWilcoxDegs_q1q4_100Genes.xlsx"))

```


## 7.5 Candidate gene lists
### 7.5.1 Gene sets
```{r}
mergedFamWilcoxDegsCTVT <- famWilcox %>%
  filter(p_val_adj < 0.05 & avg_log2FC > 2.5 & cluster=="CTVT") %>%
  filter(gene %in% q1Label100$Symbol) %>%
  dplyr::rename("Symbol"="gene") %>%
  left_join(q1Label100) %>%
  mutate("Annotation"=paste0("tumorSpec_", cluster, "Deg")) 

# 17 genes are negative correlated with tumor purity and a Bcell DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Bcells" & purity_correlation < -0)))
candBcell <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Bcells" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

bcellSet <- candBcell$Symbol

# 36 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Myeloid" & purity_correlation < -0)))
candMyeloid <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Myeloid" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

myeloidSet <- candMyeloid$Symbol

# 32 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "CAF.Endo" & purity_correlation < -0)))
candCAFEndo <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "CAF.Endo" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

cafendoSet <- candCAFEndo$Symbol


# 21 genes are negatively correlated with tumor purity and a Plasma DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Plasma" & purity_correlation < -0)))
candPlasma <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Plasma" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

plasmaSet <- candPlasma$Symbol

# 9 genes are negatively correlated with tumor purity and a Tcells DEG
print(nrow(mergedFamWilcoxDegsAdrian %>% filter(cluster == "Tcells" & purity_correlation < -0)))
candTcells <- mergedFamWilcoxDegsAdrian %>% 
  filter(cluster == "Tcells" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

tcellSet <- candTcells$Symbol
```

### 7.5.2 Upset
```{r}
geneSets <- list("ctvt"=ctvtSet,
             "bcell"=bcellSet,
             "myeloid"=myeloidSet,
             "caf.endo"=cafendoSet,
             "plasma"=plasmaSet,
             "tcell"=tcellSet)


genes <- sort(unique(unlist(geneSets)))

sets <- t(+sapply(geneSets, "%in%", x = genes)) %>%
  t() %>%
  as.data.frame()

rownames(sets) <- genes

sets <- sets %>%
  rownames_to_column("Gene")

upsetPlot <- upset(sets,
      sets = c("ctvt", "bcell", "myeloid", "caf.endo", "plasma", "tcell"),
      # keep.order = TRUE,
      nintersects = NA, # show all intersections
      order.by = "freq") # optional, could be degree or freq

upsetPlot <- upset(
  sets,
  sets = c("ctvt", "caf.endo", "myeloid", "plasma", "tcell", "bcell"),
  keep.order = TRUE,
  nintersects = NA,
  sets.bar.color = c(
    ctvt = "#E41A1C",
    caf.endo = "gold",
    myeloid = "#4DAF4A",
    plasma = "skyblue",
    tcell = "cornflowerblue",
    bcell = "pink"
  )
)


uniqueGenes <- sets %>%
  column_to_rownames("Gene") %>%
  mutate(rowsum = rowSums(.)) %>%
  filter(rowsum==1)


upsetData <- sets %>%
  unite(col = "intersection", -c("Gene"), sep = "") %>%
  group_by(intersection) %>%
  summarise(list = list(Gene)) %>%
  mutate(list = setNames(list, intersection)) %>%
  pull(list)

write.xlsx(sets, paste0(resDir, fprefix, "_geneCorrelations_famDegs.xlsx"))

pdf(paste0(plotDir, "/upset/", fprefix, "_q1q4Genes_famDegs_intersection.pdf"), 
    height=6, 
    width=8)
print(upsetPlot)
dev.off()
```

### 7.5.3 Unique gene sets
```{r}
uniqueGenesFinal <- uniqueGenes %>%
  dplyr::select(-rowsum)

# Initialize a list to store unique genes per cell type
uniqueGeneLists <- list()

# Loop through each column (cell type)
for (cell_type in colnames(uniqueGenesFinal)) {
  # Identify genes that are only expressed (1) in this cell type
  unique_genes <- rownames(uniqueGenesFinal)[
    uniqueGenesFinal[, cell_type] == 1 & rowSums(uniqueGenesFinal) == 1
  ]
  
  # Store in the list
  uniqueGeneLists[[cell_type]] <- unique_genes
}
```








# 8 eT/eH
## 8.1 Data
New approach for marker gene selection: 
- Group by gene, take median eT and eH --> probably more complete with purity corr. by gene as well. 
- Of note, max eT = Inf --> replacing with 60000. 
```{r}
# snpDf <- 
geneDf <- adrianDf %>%
  group_by( GeneID, Symbol) %>% # Group by SNP (Location)
  mutate(etEdit = ifelse(eT == Inf, 
                             50000,
                             eT)) %>%
  summarise(sampleCount = n(), 
            medianEh = median(eH), 
            medianEt = median(eT),
            medianEtEdit = median(etEdit), 
            sumEt = sum(eT),
            sumEtEdit = sum(etEdit),
            sumEh = sum(eH)) %>%
  left_join(adrianCorr, 
            by = join_by(Symbol))
```

## 8.2 Plot
### 8.2.1 eT/eH vs eT+eH
```{r}
plotDf <- geneDf %>%
  mutate(Sum = medianEh + medianEt, 
         Frac = medianEt/medianEh,
         fracEdit = medianEtEdit/medianEh,
         sumEdit = medianEh + medianEtEdit,
         # medianEtLog = log1p(medianEt),
         # medianEtEditLog = log10(medianEtEdit),
         labelTH = ifelse(medianEh > 1000 & medianEt > 1000,
                            Symbol, 
                            NA),
         labelNeither = ifelse(medianEh < 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelH =  ifelse(medianEh > 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelT = ifelse(medianEh < 1000 & medianEt > 1000,
                            Symbol, 
                            NA))

medianPlot <- ggplot(plotDf, aes(x = Sum, y =Frac)) +
  geom_point() +
  theme_minimal() +
  labs(title = "eT vs eH",
    x = "Total gene expression (eT + eH)", 
    y = "Gene expression ratio (eT/eH)"
  ) +
  scale_x_log10(guide = "axis_logticks", limits = c(-10^2, 10^4)) +
  scale_y_log10(guide = "axis_logticks", limits = c(-10^4, 10^4)) +
  geom_label_repel(
    aes(label = labelTH),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  # geom_label_repel(
  #   aes(label = labelNeither),
  #   nudge_x = 0.1, 
  #   nudge_y = 0.1,
  #   label.padding = unit(0.15, "lines"),
  #   label.size = 0.1,
  #   color = "black",
  #   fill = "firebrick2",
  #   min.segment.length = .5,
  #   segment.color = "firebrick2"  # line color from point to label
  # ) +
  
  geom_label_repel(
    aes(label = labelT),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "skyblue",
    min.segment.length = .5,
    segment.color = "skyblue"  # line color from point to label
  ) +
  geom_label_repel(
    aes(label = labelH),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "gold",
    min.segment.length = .5,
    segment.color = "gold"  # line color from point to label
  )

pdf(paste0(plotDir, "/scatter/", fprefix, "_medianEtEh_labelled.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```

#### 8.2.1.1 Logscale
```{r}
medianPlot <- ggplot(plotDf, aes(x = medianEh, y = medianEt)) +
  geom_point() +
  scale_x_log10(guide = "axis_logticks", limits=c(10^-3, 10^4)) +
  scale_y_log10(guide = "axis_logticks", limits=c(10^-3, 10^4)) + 
  theme_minimal() +
  labs(title = "eT vs eH",
    x = "median(eH)", 
    y = "median(eT)"
  )

medianLabel = ifelse(medianEh > 1000 & sampleCount > 20,
                            Symbol, 
                            NA))
meanPlot <- ggplot(tempPlotDf, aes(x=sampleCount, y=meanEhEdit)) +
  geom_point() + 
  geom_label_repel(
    aes(label = meanLabel),
    nudge_x = 0.1, 
    nudge_y = 0.1,
    label.padding = unit(0.15, "lines"),
    label.size = 0.1,
    color = "black",
    fill = "#69b3a2",
    min.segment.length = .5,
    segment.color = "#69b3a2"  # line color from point to label
  ) +
  theme_minimal()

```
#### 8.2.1.2 Absolute

### 8.2.2 Distribution of delta eT and eH vs purity correlation
#### 8.2.2.1 Delta eT and eH
```{r}
plotDf <- geneDf %>%
  mutate(deltaEtEh = medianEt-medianEh,
         medianEhLog = log1p(medianEh), 
         medianEtLog = log1p(medianEt),
         medianEtEditLog = log10(medianEtEdit),
         labelTH = ifelse(medianEh > 1000 & medianEt > 1000,
                            Symbol, 
                            NA),
         labelNeither = ifelse(medianEh < 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelH =  ifelse(medianEh > 1000 & medianEt < 1000,
                            Symbol, 
                            NA),
         labelT = ifelse(medianEh < 1000 & medianEt > 1000,
                            Symbol, 
                            NA))
```
#### 8.2.2.2 Purity vs delta
```{r}
tempPlotDf <- plotDf %>%
  rename("sampleCount"="variantCount") %>%
  mutate(
    fillPos = medianEtEdit,
    fillNeg = medianEh
  )

# Max for scaling
q1 <- subset(tempPlotDf, purity_correlation > 0 & deltaEtEh > 0)
maxq1 <- max(q1$deltaEtEh)
# q1Label <- 

q2 <- subset(tempPlotDf, purity_correlation > 0 & deltaEtEh < 0)
maxq2 <- min(q2$deltaEtEh)

q3 <- subset(tempPlotDf, purity_correlation < 0 & deltaEtEh > 0)
maxq3 <- min(q3$deltaEtEh)

q4 <- subset(tempPlotDf, purity_correlation < 0 & deltaEtEh < 0)
maxq4 <- min(q4$deltaEtEh)


medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = deltaEtEh)) +
  # NEUTRAL points layer
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = sampleCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # POS POS correlation points layer
  new_scale_fill() +
  geom_point(data = q1,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "mistyrose", high = "maroon",
    limits = c(0, 5000), na.value = "grey90"
  ) +
   # NEG POS correlation points layer
    new_scale_fill() +
  geom_point(data = q2,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    high = "#ffd580", low = "firebrick2",
    limits = c(maxq2, 0), na.value = "grey90"
  ) +
  
  # NEG POS correlation points layer
    new_scale_fill() +
  geom_point(data = q3,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "#7ED6B7", high = "#1831F5",
    limits = c(0,  5000), na.value = "grey90"
  ) +
    # NEG POS correlation points layer
  new_scale_fill() +
  geom_point(data = q4,
             aes(fill = deltaEtEh, size = sampleCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(
    name = "",
    low = "#7E53E6", high = "#D8BFD8",
    limits = c(maxq4, 0), na.value = "grey90"
  ) +
  theme_minimal() +
  labs(
    title = "",
    x = "Purity Correlation", 
    y = "Median eT - Median eH"
  ) 

pdf(paste0(plotDir, "/volcano/", fprefix, "_deltaEtEh_purity.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```
#### 8.2.2.3 Labeled purity vs delta
```{r}
# Filter points with abs(deltaEtEh) > 500 in Q1 and Q3
q1Label <- subset(q1, abs(deltaEtEh) > 200)
q4Label <- subset(q4, abs(deltaEtEh) > 200)

# Plot with labels
medianPlot <- ggplot(tempPlotDf, aes(x = purity_correlation, y = deltaEtEh)) +
  # NEUTRAL points
  geom_point(data = subset(tempPlotDf, purity_correlation >= -0.5 & purity_correlation <= 0.5),
             aes(size = variantCount), fill = "grey80", color = "black", shape = 21, stroke = 0.2) +

  # Q1
  new_scale_fill() +
  geom_point(data = q1,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "mistyrose", high = "maroon", limits = c(0, 5000), na.value = "grey90") +

  # Q2
  new_scale_fill() +
  geom_point(data = q2,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(high = "#ffd580", low = "firebrick2", limits = c(maxq2, 0), na.value = "grey90") +

  # Q3
  new_scale_fill() +
  geom_point(data = q3,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "#7ED6B7", high = "#1831F5", limits = c(0, 5000), na.value = "grey90") +

  # Q4
  new_scale_fill() +
  geom_point(data = q4,
             aes(fill = deltaEtEh, size = variantCount), color = "black", shape = 21, stroke = 0.2) +
  scale_fill_gradient(low = "#7E53E6", high = "#D8BFD8", limits = c(maxq4, 0), na.value = "grey90") +

  # Labels for Q1 and Q4
  geom_text(data = q1Label, aes(label = Symbol), size = 3, vjust = -1, check_overlap = TRUE) +
  geom_text(data = q4Label, aes(label = Symbol), size = 3, vjust = -1, check_overlap = TRUE) +
  scale_size(range = c(1,10)) +

  theme_minimal() +
  labs(x = "Purity Correlation", y = "Median eT - Median eH")

pdf(paste0(plotDir, "/volcano/", fprefix, "_deltaEtEh_purity_labelled200.pdf"), height=10, width=10)
print(medianPlot)
dev.off()
```

## 8.3 With marker genes
### 8.3.1 Candidate gene lists
```{r}
# Filter points with abs(deltaEtEh) > 500 in Q1 and Q4

# q1 = tumor
q1Label100 <- subset(q1, abs(deltaEtEh) > 100)

# q4 = host
q4Label100 <- subset(q4, abs(deltaEtEh) > 100)
```

### 8.3.2 CTVT
#### 8.3.2.1 Filter genes 
Testing positive genes from FAM with high purity correlations. 
```{r}
famWilcox <- read.xlsx(paste0(resDir, "findAllMarkers/", fprefix, "_seuratObjNoBad_famWilcox.xlsx"))

famWilcoxDegs <- famWilcox %>%
  filter(p_val_adj < 0.05 & avg_log2FC > 2.5 & cluster == "CTVT") 

intersecting <- intersect(famWilcoxDegs$gene, q1Label100$Symbol) ### 93 genes intersecting

# intersectingStrict <- intersect(famWilcoxDegs$gene, tumorSpecDfStrict$Symbol) ### None overlap. 
```

#### 8.3.2.2 Merge with purity correlations and eT
```{r}
mergedFamWilcoxDegsCTVT <- famWilcox %>%
  filter(p_val_adj < 0.01 & avg_log2FC > 2.5 & cluster=="CTVT") %>%
  filter(gene %in% q1Label100$Symbol) %>%
  dplyr::rename("Symbol"="gene") %>%
  left_join(q1Label100) %>%
  mutate("Annotation"=paste0("q1Label100_CTVTDeg")) 

write.xlsx(mergedFamWilcoxDegsCTVT,
           paste0(resDir, fprefix, "_mergedFamWilcoxDegs_q1100TumorGenes.xlsx"))

ctvtSet <- mergedFamWilcoxDegsCTVT$Symbol
```
#### 8.3.2.3 Feature plots
Merged and split by sample
```{r}
for(g in intersecting){

  feat <- FeaturePlot(seuratObjNoBad,
                      reduction = "TSNE_log_corrected",
                      features=g) +

    ggtitle(paste0("q1 + Wilcox DEG - ", g))

  featSamp <- FeaturePlot(seuratObjNoBad,
                          reduction = "TSNE_log_corrected",
                          features=g,
                          split.by = "Sample")

  pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_q1label100_famWilcoxDegs_", g,"_expression.pdf"),
      height=4,
      width=4)

  print(feat)

  dev.off()



  pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix,  "/featureSeurat/tumorSpecificIntersect/03_Deconvolution_CTVT_q1label100_famWilcoxDegs_", g,"_expressionBySample.pdf"),
      height=4,
      width=14)

  print(featSamp)

  dev.off()

}
```

### 8.3.2 Host specific genes
#### 8.3.2.1 Feature plots
```{r}
###### 2.1.4 Loop through NON CTVT FAM DEG's --> check against genetic deconv list. Test positive genes first.
for(ctype in unique(famWilcox$cluster)){
  if(ctype=="CTVT"){
    next
  }

  tempFamWilcoxDegs <- famWilcox %>%
    filter(p_val_adj < 0.05 & avg_log2FC > 2.5 & cluster == ctype)

  tempIntersecting <- intersect(tempFamWilcoxDegs$gene, q4Label100$Symbol)

  print(paste0("Cell type: ", ctype, " - Intersecting: ", length(tempIntersecting)))

  tempPlotDir <- paste0("hostSpecificIntersect", ctype)

  for(tg in tempIntersecting){

    feat <- FeaturePlot(seuratObjNoBad,
                        reduction = "TSNE_log_corrected",
                        features=tg) +

      ggtitle(paste0("q4Label100 + ", ctype, " Wilcox DEG - ", tg))

    featSamp <- FeaturePlot(seuratObjNoBad,
                            reduction = "TSNE_log_corrected",
                            features=tg,
                            split.by = "Sample")

    pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_q4Label100_famWilcoxDegs_", tg,"_expression.pdf"),
        height=4,
        width=4)

    print(feat)

    dev.off()



    pdf(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/034_Plots/", fprefix, "/featureSeurat/", tempPlotDir, "/03_Deconvolution_CTVT_q4Label100_famWilcoxDegs_", tg,"_expressionBySample.pdf"),
        height=4,
        width=14)

    print(featSamp)

    dev.off()

  }

}
```

### 6.4.2 Merge with purity correlations
```{r}
# mergedFamWilcoxDegsAdrian <- data.frame()
# 
# 
# for(ctype in unique(famWilcox$cluster)){
#   if(ctype=="CTVT"){
#     next
#   }
#   
#   temp <- famWilcox %>%
#     filter(p_val_adj < 0.05 & avg_log2FC > 2.5& cluster==ctype) %>%
#     filter(gene %in% q4Label100$Symbol) %>%
#     dplyr::rename("Symbol"="gene") %>%
#     left_join(q4Label100) %>%
#     mutate("Annotation"=paste0("q4Label100_", ctype, "Deg"))
#   
#   mergedFamWilcoxDegsAdrian <- rbind(mergedFamWilcoxDegsAdrian, temp)
#   
# }
# 
# allDegs <- rbind(mergedFamWilcoxDegsAdrian, mergedFamWilcoxDegsCTVT)
# 
# write.xlsx(allDegs,
#            paste0(resDir, fprefix, "_mergedFamWilcoxDegs_q1q4_100Genes.xlsx"))


allDegs <- read.xlsx(paste0(resDir, fprefix, "_mergedFamWilcoxDegs_q1q4_100Genes.xlsx"))

```


## 8.5 Candidate gene lists
### 8.5.1 Gene sets
Includes marker genes shared by different clusters. 
```{r}
print(nrow(allDegs %>% filter(cluster == "CTVT" & purity_correlation > 0)))
candCTVT <- allDegs %>% 
  filter(cluster == "CTVT" & purity_correlation > -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

ctvtSet <- candCTVT$Symbol

# 17 genes are negative correlated with tumor purity and a Bcell DEG
print(nrow(allDegs %>% filter(cluster == "Bcells" & purity_correlation < 0)))
candBcell <- allDegs %>% 
  filter(cluster == "Bcells" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

bcellSet <- candBcell$Symbol

# 36 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(allDegs %>% filter(cluster == "Myeloid" & purity_correlation < 0)))
candMyeloid <- allDegs %>% 
  filter(cluster == "Myeloid" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

myeloidSet <- candMyeloid$Symbol

# 32 genes are negatively correlated with tumor purity and a Myeloid DEG
print(nrow(allDegs %>% filter(cluster == "CAF.Endo" & purity_correlation < 0)))
candCAFEndo <- allDegs %>% 
  filter(cluster == "CAF.Endo" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

cafendoSet <- candCAFEndo$Symbol


# 21 genes are negatively correlated with tumor purity and a Plasma DEG
print(nrow(allDegs %>% filter(cluster == "Plasma" & purity_correlation < 0)))
candPlasma <- allDegs %>% 
  filter(cluster == "Plasma" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

plasmaSet <- candPlasma$Symbol

# 9 genes are negatively correlated with tumor purity and a Tcells DEG
print(nrow(allDegs %>% filter(cluster == "Tcells" & purity_correlation < 0)))
candTcells <- allDegs %>% 
  filter(cluster == "Tcells" & purity_correlation < -0) %>% 
  dplyr::select(Symbol, medianEh, purity_correlation, avg_log2FC)

tcellSet <- candTcells$Symbol
```

### 8.5.2 Upset
```{r}
geneSets <- list("ctvt"=ctvtSet,
             "bcell"=bcellSet,
             "myeloid"=myeloidSet,
             "caf.endo"=cafendoSet,
             "plasma"=plasmaSet,
             "tcell"=tcellSet)


genes <- sort(unique(unlist(geneSets)))

sets <- t(+sapply(geneSets, "%in%", x = genes)) %>%
  t() %>%
  as.data.frame()

rownames(sets) <- genes

sets <- sets %>%
  rownames_to_column("Gene")

upsetPlot <- upset(sets,
      sets = c("ctvt", "bcell", "myeloid", "caf.endo", "plasma", "tcell"),
      # keep.order = TRUE,
      nintersects = NA, # show all intersections
      order.by = "freq") # optional, could be degree or freq

upsetPlot <- upset(
  sets,
  sets = c("ctvt", "caf.endo", "myeloid", "plasma", "tcell", "bcell"),
  keep.order = TRUE,
  nintersects = NA,
  sets.bar.color = c(
    ctvt = "#E41A1C",
    caf.endo = "gold",
    myeloid = "#4DAF4A",
    plasma = "skyblue",
    tcell = "cornflowerblue",
    bcell = "pink"
  )
)


uniqueGenes <- sets %>%
  column_to_rownames("Gene") %>%
  mutate(rowsum = rowSums(.)) %>%
  filter(rowsum==1)


upsetData <- sets %>%
  unite(col = "intersection", -c("Gene"), sep = "") %>%
  group_by(intersection) %>%
  summarise(list = list(Gene)) %>%
  mutate(list = setNames(list, intersection)) %>%
  pull(list)

write.xlsx(sets, paste0(resDir, fprefix, "_geneCorrelations_famDegs.xlsx"))

pdf(paste0(plotDir, "/upset/", fprefix, "_q1q4Genes_famDegs_intersection.pdf"), 
    height=6, 
    width=8)
print(upsetPlot)
dev.off()
```

### 8.5.3 Unique gene sets
```{r}
uniqueGenesFinal <- uniqueGenes %>%
  dplyr::select(-rowsum)

# Initialize a list to store unique genes per cell type
uniqueGeneLists <- list()

# Loop through each column (cell type)
for (cell_type in colnames(uniqueGenesFinal)) {
  # Identify genes that are only expressed (1) in this cell type
  unique_genes <- rownames(uniqueGenesFinal)[
    uniqueGenesFinal[, cell_type] == 1 & rowSums(uniqueGenesFinal) == 1
  ]
  
  # Store in the list
  uniqueGeneLists[[cell_type]] <- unique_genes
}

# allDegs$Unique <- ifelse(allDegs$Symbol %in% rownames(uniqueGenesFinal), "Unique", "Shared")


### Create master table to merge with curated gene list (see next chunk).

masterMarkerGenes <- allDegs %>%
  mutate(Unique=ifelse(Symbol %in% rownames(uniqueGenesFinal), "Unique", "Shared"), 
         Type = ifelse(cluster == "CTVT", 
                       "100q1Wilcox", 
                       "100q4Wilcox"),
         
         corrQuad = ifelse(cluster == "CTVT", 
                       "q1", 
                       "q4"))
```


### 8.5.4 Manually curate
Get famWilcox and gene correlation information for curated (human) gene list. Subject to changing. 
```{r}
immune_set <- list(
  `Bcells` = c("MS4A1", "PAX5", "BANK1", "BCL11A"),
  `Tcells` = c("CD3E", "CD3G", "CD4", "CD8A", "ICOS", "CTLA4"),
  # `MHCI` = c("TAP1", "B2M", "DLA-88", "DLA-64"),
  # `MHCII`= c("DLA-DRA", "DLA-DQA1", "DLA-DMA"),
  `Plasma` = c("JCHAIN", "IRF4", "POU2AF1"),
  `Myeloid` = c("LYZ", "CD80", "CD86", "IL18", "MRC1", "GAB2"),
  `CAF.Endo`= c("FAP", "CALD1", "COL6A3", "LAMA4", "ADGRL4", "VWF", "PDGFD"),
  `CTVT` = c("LSAMP", "DPP10", "PDE1C", "PDGFC", "DLC1", "EPO", "MS4A2", "KIT") 
)

for (set in 1:length(immune_set)){
  s <- immune_set[set]
  tempName <- names(s)
  
  ### Filter genes from list not assigned to cluster of interest in find all markers results. Same for genes not included in Adrian's table?
  
  tempWilcox <- famWilcox %>%
    filter(cluster==tempName & gene %in% s[[1]] & gene %in% tempPlotDf$Symbol)
  
  print(paste0(setdiff(s[[1]], tempWilcox$gene), " excluded from manual list because not found in FAM results for ", tempName, "."))
  
  print(paste0(setdiff(s[[1]], tempPlotDf$Symbol), " excluded from manual list because not found in Adrian's table. "))
  
  tempCorr <- tempPlotDf %>% 
    filter(Symbol %in% tempWilcox$gene) %>%
    mutate(corrQuad = ifelse(Symbol %in% q1$Symbol, "q1",
                             ifelse(Symbol %in% q2$Symbol, "q2", ifelse(Symbol %in% q3$Symbol, "q3", "q4"))))
  
  tempDf <- cbind(tempWilcox, 
                  tempCorr) %>%
    mutate(Unique = NA,
           
           Annotation =paste0(corrQuad, "Manual_", tempName, 
                         ifelse(p_val_adj<0.05 & avg_log2FC >2.5, "Deg", "")),
           Type = paste0(corrQuad, "Manual"))
  
  masterMarkerGenes <- rbind(masterMarkerGenes, 
                             tempDf[,colnames(masterMarkerGenes)])
}

### Update shared column!
table(masterMarkerGenes$cluster)


```

### 8.5.5 Update unique gene sets
```{r}
geneSets <- list("ctvt"=masterMarkerGenes[which(masterMarkerGenes$cluster=="CTVT") , ]$Symbol,
             "bcell"=masterMarkerGenes[which(masterMarkerGenes$cluster=="Bcells") , ]$Symbol,
             "myeloid"=masterMarkerGenes[which(masterMarkerGenes$cluster=="Myeloid") , ]$Symbol,
             "caf.endo"=masterMarkerGenes[which(masterMarkerGenes$cluster=="CAF.Endo") , ]$Symbol,
             "plasma"=masterMarkerGenes[which(masterMarkerGenes$cluster=="Plasma") , ]$Symbol,
             "tcell"=masterMarkerGenes[which(masterMarkerGenes$cluster=="Tcells") , ]$Symbol)


genes <- sort(unique(unlist(geneSets)))

sets <- t(+sapply(geneSets, "%in%", x = genes)) %>%
  t() %>%
  as.data.frame()

rownames(sets) <- genes

sets <- sets %>%
  rownames_to_column("Gene")

upsetPlot <- upset(sets,
      sets = c("ctvt", "bcell", "myeloid", "caf.endo", "plasma", "tcell"),
      # keep.order = TRUE,
      nintersects = NA, # show all intersections
      order.by = "freq") # optional, could be degree or freq

upsetPlot <- upset(
  sets,
  sets = c("ctvt", "caf.endo", "myeloid", "plasma", "tcell", "bcell"),
  keep.order = TRUE,
  nintersects = NA,
  sets.bar.color = c(
    ctvt = "#E41A1C",
    caf.endo = "gold",
    myeloid = "#4DAF4A",
    plasma = "skyblue",
    tcell = "cornflowerblue",
    bcell = "pink"
  )
)


uniqueGenes <- sets %>%
  column_to_rownames("Gene") %>%
  mutate(rowsum = rowSums(.)) %>%
  filter(rowsum==1)


upsetData <- sets %>%
  unite(col = "intersection", -c("Gene"), sep = "") %>%
  group_by(intersection) %>%
  summarise(list = list(Gene)) %>%
  mutate(list = setNames(list, intersection)) %>%
  pull(list)

write.xlsx(sets, paste0(resDir, fprefix, "_masterMarkerGenes.xlsx"))

pdf(paste0(plotDir, "/upset/", fprefix, "_masterMarkerGenes_intersection.pdf"), 
    height=6, 
    width=8)
print(upsetPlot)
dev.off()


uniqueGenesFinal <- uniqueGenes %>%
  dplyr::select(-rowsum)

# Initialize a list to store unique genes per cell type
uniqueGeneLists <- list()

# Loop through each column (cell type)
for (cell_type in colnames(uniqueGenesFinal)) {
  # Identify genes that are only expressed (1) in this cell type
  unique_genes <- rownames(uniqueGenesFinal)[
    uniqueGenesFinal[, cell_type] == 1 & rowSums(uniqueGenesFinal) == 1
  ]
  
  # Store in the list
  uniqueGeneLists[[cell_type]] <- unique_genes
}

masterMarkerGenes <- masterMarkerGenes %>%
  mutate(Unique=ifelse(Symbol %in% rownames(uniqueGenesFinal), "Unique", "Shared"))

### None of the master marker genes added by manual are overlapping. 

write.xlsx(masterMarkerGenes, paste0(resDir, fprefix, "_masterMarkerGenes.xlsx"))

table(masterMarkerGenes$Symbol) # Some from manual overlap with fam/corr degs. 
```



