---
title: "01_Deconvolution_reDeconv"
author: "IMEV2"
date: "`r Sys.Date()`"
output: html_document
---

# 0 Environment
## 0.1 Packages
### 0.1.1 Installs
```{r echo=F, message=F, warning=FALSE}
# if (!requireNamespace("devtools", quietly = TRUE)) {
#    install.packages("devtools")
# }
# devtools::install_github("Danko-Lab/BayesPrism/BayesPrism")
```
### 0.1.1 Library
```{r echo=F, message=F, warning=FALSE}
set.seed(123)
library(tidyverse)
library(SingleCellExperiment)
library(openxlsx)
library(patchwork)
library(pheatmap)
library(ComplexHeatmap)
library(circlize)
setClassUnion("ExpData", c("matrix", "SummarizedExperiment"))
```

## 0.2 Paths
```{r echo=F, message=FALSE, warning=FALSE}
fprefix <- "03_Deconvolution_CTVT_reDeconv"
wDir <- paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/")
dataDir <- paste0(wDir, "/031_Data/")
resDir <- paste0(wDir, "/033_Results/", fprefix, "/")
plotDir <- paste0(wDir, "/034_Plots/", fprefix, "/")

dir.create(resDir)
dir.create(plotDir)

scriptDir <- paste0(wDir, "/032_Scripts/03_Deconvolution_CTVT_deconBenchmark/")
source(paste0(scriptDir, "03_Deconvolution_CTVT_deconBenchmark_metrics.R"))
```

# 1.0 Inputs
BayesPrism will take intersect of genes in reference and bulk samples.
Use unnormalized and untransformed count data.

## 1.1 Genes of interest
### 1.1.1 All 
Mixed curated and non-curated list. 
```{r}
masterMarkerGenes <- read.xlsx("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_masterMarkerGenes.xlsx")
  
geneDict <- masterMarkerGenes %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

uniqueGeneDict <- masterMarkerGenes %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

table(masterMarkerGenes$cluster)

cibersortMin50Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr50_minGenes2to10/CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")

cibersortDictMin50 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin50Genes2to10$NAME) %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

cibersortUniqueDictMin50 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin50Genes2to10$NAME & Unique == "Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()


cibersortMin10Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr10_minGenes2to10/CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")

cibersortDictMin10 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin10Genes2to10$NAME) %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

cibersortUniqueDictMin10 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin10Genes2to10$NAME & Unique == "Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()



```

### 1.1.1 All 
```{r}

resTypeMatched <- read.xlsx(paste0(wDir,"/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resTypeMatched.xlsx"))

resTypeMatchedDegs <- resTypeMatched %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
  rownames_to_column("GeneID") 

curated <- masterMarkerGenes %>%
  filter(GeneID != intersect(resTypeMatchedDegs$GeneID, masterMarkerGenes$GeneID))

curatedCib10 <- cibersortDictMin10 

curatedCib10Unique <- cibersortUniqueDictMin10

intersect(resTypeMatchedDegs$GeneID, cibersortDictMin50$GeneID)

curatedCib50 <- cibersortDictMin50 %>%
  filter(GeneID != intersect(resTypeMatchedDegs$GeneID, cibersortDictMin50$GeneID))

curatedCib50Unique <- cibersortUniqueDictMin50 %>%
  filter(GeneID != intersect(resTypeMatchedDegs$GeneID, cibersortUniqueDictMin50$GeneID))
```



## 1.2 Reference
Reference - cellxgene dense matrix.
### 1.2.1 Matrix
```{r}
obj <- readRDS(paste0(dataDir, "SCE_concat_CTVT_includingAllGenes_logCorrecDone_vstCorrecDone_ClustDone2_ManAnot.rds"))

objNoBad <- obj[,obj$manual.log.anot.fine!="bad"]

rm(obj)
gc()



snCountsFinal <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  # filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 

snCountsFinalMarkers <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 

snCountsFinalMarkersUnique <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 


snCountsFinalCib10 <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 

snCountsFinalCib10Unique <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 


snCountsFinalCib50 <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 

snCountsFinalCib50Unique <- counts(objNoBad) %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol") 

colData <- colData(objNoBad)
# rm(objNoBad)
gc()

# Create the cell count matrix: rows = cell types, columns = samples
snCellCountsNoBad <- colData(objNoBad) %>%
  as.data.frame() %>%
  group_by(manual.log.anot.fine,
           Sample) %>%
  summarise(cellCount = n()) %>%
  pivot_wider(names_from = Sample,
              values_from = cellCount,
              values_fill = 0) %>%
  column_to_rownames("manual.log.anot.fine")

snCellPropNoBad <- apply(snCellCountsNoBad, 2, function(x) x/sum(x))

colSums(snCellPropNoBad) # Should all be 1.
```

## 1.3 Bulk
### 1.2.1 Matrix
```{r}
rowData <- rowData(objNoBad) %>%
  as.data.frame() %>%
  rename("ID"="GeneID")

# bulkCountsFinal <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
#   as.data.frame() %>%
#   # select gene id's (stored as columns in df) that are hvgs.
#   rownames_to_column("GeneID") %>%
#   left_join(rowData) 

# bulkCountsFinal <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds"), 
#                                                        unit="TPM")) %>%
#   as.data.frame() %>%
#   # select gene id's (stored as columns in df) that are hvgs.
#   rownames_to_column("GeneID") %>%
#   left_join(rowData) %>%
#   select(-c(S))


duplicatedSymbols <- table(bulkCountsFinal$Symbol) %>% 
  as.data.frame() %>%
  filter(Freq > 1)

uniqueSymbols <- table(bulkCountsFinal$Symbol) %>% 
  as.data.frame() %>%
  filter(Freq == 1)

bulkNonUnique <- bulkCountsFinal %>%
  filter(Symbol %in% duplicatedSymbols$Var1)

duplicatedIDs <- table(bulkCountsFinal$GeneID) %>% 
  as.data.frame() %>%
  filter(Freq > 1)

bulkNonUniqueIDs <- bulkCountsFinal %>%
  filter(GeneID %in% duplicatedIDs$Var1)

bulkCountsFinal <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  left_join(rowData) %>%
  filter(Symbol %in% uniqueSymbols$Var1) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 

bulkCountsFinalMarkers <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% geneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 

bulkCountsFinalMarkersUnique <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% uniqueGeneDict$GeneID) %>%
  left_join(uniqueGeneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 

bulkCountsFinalCib10 <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% cibersortDictMin10$GeneID) %>%
  left_join(cibersortDictMin10) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 

bulkCountsFinalCib10Unique <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% cibersortUniqueDictMin10$GeneID) %>%
  left_join(cibersortUniqueDictMin10) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 


bulkCountsFinalCib50 <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% cibersortDictMin50$GeneID) %>%
  left_join(cibersortDictMin50) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)


bulkCountsFinalCib50Unique <- readRDS(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  # select gene id's (stored as columns in df) that are hvgs.
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% cibersortUniqueDictMin50$GeneID) %>%
  left_join(cibersortUniqueDictMin50) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID) 


bulkCpm <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "CPM")%>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  left_join(rowData) %>%
  filter(Symbol %in% uniqueSymbols$Var1) %>%
  column_to_rownames("Symbol") %>%
  select(-c(GeneID, Type, Chromosome, hvgs) )
```
### 1.3.2 Metadata
```{r}
bulkCountsFinal_metadata <- read.xlsx(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal_metadata.xlsx"))
```
# 2 Export
## 2.1 Reference
A Tab-delimited table. The first row is for cell IDs and each of other rows is for a gene. The first column is for gene names and each of other column is for a cell, where all cell IDs in the first row should match the cell IDs in the Meta data file. The scRNA-seq data must be in raw count.
```{r}
snCountsExport <- snCountsFinal %>%
            rownames_to_column("geneSample")

write.table(snCountsExport, 
          paste0(resDir, fprefix, "_snCountsExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")

snCountsMarkersExport <- snCountsFinalMarkers %>%
            rownames_to_column("geneSample")

write.table(snCountsMarkersExport, 
          paste0(resDir, fprefix, "_snCountsMarkersExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


snCountsMarkersUniqueExport <- snCountsFinalMarkersUnique %>%
            rownames_to_column("geneSample")

write.table(snCountsMarkersUniqueExport, 
          paste0(resDir, fprefix, "_snCountsMarkersUniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")



snCountsCib10Export <- snCountsFinalCib10 %>%
            rownames_to_column("geneSample")

write.table(snCountsCib10Export, 
          paste0(resDir, fprefix, "_snCountsCib10Export.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


snCountsCib10UniqueExport <- snCountsFinalCib10Unique %>%
            rownames_to_column("geneSample")

write.table(snCountsCib10UniqueExport, 
          paste0(resDir, fprefix, "_snCountsCib10UniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")



snCountsCib50Export <- snCountsFinalCib50 %>%
            rownames_to_column("geneSample")

write.table(snCountsCib50Export, 
          paste0(resDir, fprefix, "_snCountsCib50Export.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


snCountsCib50UniqueExport <- snCountsFinalCib50Unique %>%
            rownames_to_column("geneSample")

write.table(snCountsCib50UniqueExport, 
          paste0(resDir, fprefix, "_snCountsCib50UniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")
```

## 2.2 Meta
A Tab-delimited table. The first row is for column names and each of other rows is for a cell. The first column of the table is for cell IDs. The second column is for cell types and the third column is for sample IDs. The table can have more columns. However, the program only uses the first three columns.
```{r}
metaExport <- colData %>%
  as.data.frame() %>%
  rownames_to_column("cellId") %>%
  select(c(cellId, manual.log.anot.fine, Sample)) %>%
  rename("manual.log.anot.fine"="cellType")

write.table(metaExport, 
          paste0(resDir, fprefix, "_metaExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")
```

## 2.3 Bulk
A Tab-delimited table for the expression information (RPK/TPM/RPKM) of mixture samples. The first row is for sample IDs and each of other rows is for a gene. The first column is for gene names and each of other columns is for a mixture sample. The bulk RNA-seq data and the signature gene matrix should have enough common genes to guarantee the performance of deconvolution.
```{r}
bulkCountsExport <- bulkCountsFinal %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")

bulkCpmExport <- bulkCpm %>%
            rownames_to_column("geneSample")

write.table(bulkCpmExport, 
          paste0(resDir, fprefix, "_bulkCpmExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")

bulkCountsMarkersExport <- bulkCountsFinalMarkers %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsMarkersExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


bulkCountsMarkersUniqueExport <- bulkCountsFinalMarkersUnique %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsMarkersUniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")



bulkCountsCib10Export <- bulkCountsFinalCib10 %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsCib10Export.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


bulkCountsCib10UniqueExport <- bulkCountsFinalCib10Unique %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsCib10UniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")



bulkCountsCib50Export <- bulkCountsFinalCib50 %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsCib50Export.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")


bulkCountsCib50UniqueExport <- bulkCountsFinalCib50Unique %>%
            rownames_to_column("geneSample")

write.table(bulkCountsExport, 
          paste0(resDir, fprefix, "_bulkCountsCib50UniqueExport.txt"), 
          quote=F,
          col.names = T,
          row.names = F, 
          sep = "\t")
```


# 3 Filter sig. matrix
## 3.1 Read in sig. matrix
### 3.1.1 Normal
```{r message=FALSE, warning=FALSE}
sig <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0.tsv") 

colNames <- c("Gene", colnames(sig))

sig <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0.tsv", col_names = colNames)[-1,] %>%
  column_to_rownames("Gene") 

sigNum <- as.data.frame(apply(sig, 2, as.numeric))

rownames(sigNum) <- rownames(sig)

sigFinal <- sigNum %>%
  rownames_to_column("Symbol")
```
### 3.1.2 Filter by matched degs
```{r}
rowData <- rowData(objNoBad) %>%
  as.data.frame() %>%
  select(c(ID, Symbol))%>%
  rename("ID"="GeneID")

resTypeMatchedDegs <- read.xlsx( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resTypeMatched.xlsx") %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
  left_join(rowData)  %>%
  left_join(sigFinal)

# 2 genes from sig file are DEGs between ddsTypeMatched.
# > intersect(resTypeMatchedDegs$GeneID, sigFinal$Symbol)
# [1] "ENSCAFG00000032358" "ENSCAFG00000041010"
# length(intersect(resTypeMatchedDegs$GeneID, sigFinal$Symbol))

sigFinalFiltered <- sigFinal %>%
  filter(!Symbol %in% intersect(resTypeMatchedDegs$GeneID, sigFinal$Symbol)) %>%
  column_to_rownames("Symbol")

write.table(sigFinalFiltered, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigFinalFiltered.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```


## 3.1 Candidate gene lists
### 3.1.1 All markers
```{r message=FALSE, warning=FALSE}
sigMarkers <- sigFinal %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol")

write.table(sigMarkers, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigMarkers.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigMarkersUnique <- sigFinal %>%
  filter(Symbol %in% uniqueGeneDict$Symbol) %>%
  column_to_rownames("Symbol")

write.table(sigMarkersUnique , "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigMarkersUnique.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```

### 3.1.2 Cib 10
```{r message=FALSE, warning=FALSE}
sigCib10 <- sigFinal %>%
  filter(Symbol %in% cibersortDictMin10$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib10 , "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib10.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigCib10Unique <- sigFinal %>%
  filter(Symbol %in% cibersortUniqueDictMin10$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib10Unique, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib10Unique.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```

### 3.1.3 Cib 50
```{r message=FALSE, warning=FALSE}
sigCib50 <- sigFinal %>%
  filter(Symbol %in% cibersortDictMin50$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib50, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib50.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigCib50Unique <- sigFinal %>%
  filter(Symbol %in% cibersortUniqueDictMin50$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib50Unique, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib50Unique.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```


## 3.2 Overlap with Adrian's list only
- Because this program is essentially running a type of FAM, maybe we only need to add a purity metric?


# 3 Filter sig. matrix with shift
## 3.1 Read in sig. matrix with shift
```{r message=FALSE, warning=FALSE}
sig <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_withShift.tsv") 

colNames <- c("Gene", colnames(sig))

sig <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_withShift.tsv", col_names = colNames)[-1,] %>%
  column_to_rownames("Gene") 

sigNum <- as.data.frame(apply(sig, 2, as.numeric))

rownames(sigNum) <- rownames(sig)

sigFinal <- sigNum %>%
  rownames_to_column("Symbol")
```

## 3.1 Candidate gene lists
Manually add a " " before the first column name to match original sig. matrix from ReDeconv. 
### 3.1.1 All markers
```{r message=FALSE, warning=FALSE}
sigMarkers <- sigFinal %>%
  filter(Symbol %in% geneDict$Symbol) %>%
  column_to_rownames("Symbol")

write.table(sigMarkers, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigMarkers_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigMarkersUnique <- sigFinal %>%
  filter(Symbol %in% uniqueGeneDict$Symbol) %>%
  column_to_rownames("Symbol")

write.table(sigMarkersUnique , "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigMarkersUnique_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```

### 3.1.2 Cib 10
```{r message=FALSE, warning=FALSE}
sigCib10 <- sigFinal %>%
  filter(Symbol %in% cibersortDictMin10$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib10 , "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib10_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigCib10Unique <- sigFinal %>%
  filter(Symbol %in% cibersortUniqueDictMin10$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib10Unique, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib10Unique_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```

### 3.1.3 Cib 50
```{r message=FALSE, warning=FALSE}
sigCib50 <- sigFinal %>%
  filter(Symbol %in% cibersortDictMin50$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib50, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib50_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)

sigCib50Unique <- sigFinal %>%
  filter(Symbol %in% cibersortUniqueDictMin50$Symbol)%>%
  column_to_rownames("Symbol")


write.table(sigCib50Unique, "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/Signature_mean_std_fd2.0_custom_sigCib50Unique_withShift.tsv",
            sep = "\t",
            row.names = T,
            quote= F)
```


## 3.2 Overlap with Adrian's list only
- Because this program is essentially running a type of FAM, maybe we only need to add a purity metric?

# 4 Visualizations
```{r}
# Step 2: Define a color mapping
celltype_colors <- c(
  "CTVT" = "#BCD0C7",
  "CAF.Endo" = "#FDDFA4FF",
  "Myeloid" = "#F5FDC6",
  "Plasma" = "#B4B9E0FF",
  "Tcells" = "#D56AA0",
  "Bcells" = "#A8D5E2"
  
)

groundTruth <- snCellPropNoBad %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion") %>%
  mutate(
         Sample = paste0(Sample, " - GT"))
```

## 4.1 Bar plots
```{r}
res <- list.files(path="~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/", pattern = "ReDeconv_results", full.names = T)

names <- list.files(path="~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_reDeconv/Results_4_deconvolution/", pattern = "ReDeconv_results", full.names = F)

for (r in 1:length(res)){
  
  temp <- read_tsv(res[r]) %>%
    pivot_longer(cols = -c("sample/cell_type"),
                 values_to = "Proportion",
                 names_to = "Cluster") %>%
    rename("sample/cell_type"="Sample") %>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))
  
  tempName <- str_replace(names[r], ".tsv", "")
  
  
  temp$Cluster <- factor(temp$Cluster, levels = rev(c("CTVT", 
                                                    "Myeloid", 
                                                    "Bcells",
                                                    "CAF.Endo",
                                                    "Tcells",
                                                    "Plasma")))

  # Step 3: Plot
  plot <- ggplot(
    temp,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
    ggtitle(paste0("ReDeconv - ", tempName))
  
  
   pdf(paste0(plotDir, "/bar/", fprefix, "_", tempName, ".pdf"),
      height=4, width=7)
  print(plot)
  dev.off()
  
}
```


