---
title: "01_Deconvolution_deseq"
author: "IMEV2"
date: "`r Sys.Date()`"
output: html_document
---

# 0 Environment
## 0.1 Packages
### 0.1.1 Installs
```{r echo=F, message=F, warning=FALSE}
# if (!requireNamespace("devtools", quietly = TRUE)) {
#    install.packages("devtools")
# }
# devtools::install_github("Danko-Lab/BayesPrism/BayesPrism")
```
### 0.1.1 Library
```{r echo=F, message=F, warning=FALSE}
set.seed(123)
library(vegan)
library(GGally)
library(tidyverse)
library(tximport)
library(DESeq2)
library(apeglm)
# library(ade4)
#library(ggord)
library(UpSetR)
library(cowplot)
library(ape)
library(knitr)
library(openxlsx)
# setClassUnion("ExpData", c("matrix", "SummarizedExperiment"))
```

## 0.2 Paths
```{r echo=F, message=FALSE, warning=FALSE}
fprefix <- "03_Deconvolution_CTVT_DESEQ"
wDir <- paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/")
dataDir <- paste0(wDir, "/031_Data/")
resDir <- paste0(wDir, "/033_Results/", fprefix, "/")
plotDir <- paste0(wDir, "/034_Plots/", fprefix, "/")

dir.create(resDir)
dir.create(plotDir)
```

## 0.3 Maurine's data
```{r}
sampleTableI <- read_csv("~/Desktop/1_Deconvolution/For_Nina/01_RNAseq_analysis/02_CTVT/SampleRNAseqInfo.csv") %>%
  dplyr::rename(sampleName = sampleID) %>%
  mutate(sample = str_replace(sampleName, "_p|_n|_2", "")) %>%
  mutate(ind = str_sub(sampleName, 1, 4)) %>%
  mutate(matched = ifelse(sample %in% c("3145T11a",
                                    "3145T1c",
                                    "3203T1A2b",
                                    "3203T1A2c",
                                    "3302T1b", 
                                    "3621T"),
                      "Matched", "Unmatched")) %>%
  mutate(typeOverall = str_replace(type, "_p", "")) %>%
  mutate(fileName = paste0(sampleName, "_htseqcount.txt")) %>%
  dplyr::select(sampleName, fileName, everything()) # this order matters for DESeqDataSetFromHTSeqCount


directory <- c("~/Desktop/1_Deconvolution/For_Nina/01_RNAseq_analysis/02_CTVT/Data_input_htseq_reverse/")

col_sampl <- c(`3145T11a_n` = "#FF7F0E",
               `3145T11a` = "#FFBB78",
               `3145T1c_n` = "#D62728",
               `3145T1c_p` = "#FF9896",#polyA
               #`3145T1c` = "tomato", #not yet runned 
               `3203T1A2b_n` = "#1F77B4",
               `3203T1A2b_2` = "#AEC7E8",
               `3203T1A2c_n` = "#17BECF",
               `3203T1A2c_2` = "#9EDAE5",
               `3302T1b_n` = "#9467BD",
               `3302T1b` = "#C5B0D5",
               `3302T1b_p` = "#F7B6D6",#polyA
               `3621T_n` = "#2CA02C",
               `3621T` = "#98DF8A",
               
               `3752T` = "#BCBD22", #BG
               `3759T` = "#DBDB90", #BG
               `3760T` = "#8C564B", #A
               `3761T` = "darkolivegreen3", #BG
               `3764T` = "pink4", #A
               `3770T` = "#C7C7C7", #ABG
               `3772T` = "#C49C94", #A
               `3774T` = "salmon3", #A
               `3775T` = "darkolivegreen", #BG
               `3776T` = "olivedrab3") #BG

col_Ind <- c(`3145T11a` = "#FFBB78",
               `3145T1c` = "#FF9896",#polyA
               `3203T1A2b` = "#AEC7E8",
               `3203T1A2c` = "#9EDAE5",
               `3302T1b` = "#C5B0D5",
               `3621T` = "#98DF8A",
             
               `3752T` = "#BCBD22", #BG
               `3759T` = "#DBDB90", #BG
               `3760T` = "#8C564B", #A
               `3761T` = "darkolivegreen3", #BG
               `3764T` = "pink4", #A
               `3770T` = "#C7C7C7", #ABG
               `3772T` = "#C49C94", #A
               `3774T` = "salmon3", #A
               `3775T` = "darkolivegreen", #BG
               `3776T` = "olivedrab3")
```


# 1 Data
## 1.1 Design ~ type (filter out non-matched)
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsType <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c" &
                                                                              matched=="Matched"),
                                       directory = directory,
                                       design = ~ type)

ddsType <- estimateSizeFactors(ddsType)

# add annotation
rowData(ddsType)$ENS_gene_id <- rownames(rowData(ddsType))

normalizedCountsDdsType <- counts(ddsType, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsType) >= 10) >= smallestGroupSize
ddsType <- ddsType[keep,]
colSums(counts(ddsType)) %>% barplot

ddsType$type <- factor(ddsType$type, levels = c("bulk_nuclei", "bulk_cell", "bulk_cell_p"))
```

## 1.2 Design ~ typeOverall (filter out matched)
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsTypeOverall <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c"&
                                                                                     matched == "Matched"),
                                       directory = directory,
                                       design = ~ typeOverall)

ddsTypeOverall <- estimateSizeFactors(ddsTypeOverall)

# add annotation
rowData(ddsTypeOverall)$ENS_gene_id <- rownames(rowData(ddsTypeOverall))

normalizedCountsDdsTypeOverall <- counts(ddsTypeOverall, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsTypeOverall) >= 10) >= smallestGroupSize
ddsTypeOverall <- ddsTypeOverall[keep,]
colSums(counts(ddsTypeOverall)) %>% barplot

# ddsTypeOverall$matched <- factor(ddsTypeOverall$matched, levels = c("Matched", "Unmatched"))
ddsTypeOverall$typeOverall <- factor(ddsTypeOverall$typeOverall, levels = c("bulk_nuclei", "bulk_cell"))
```



## 1.2 Design ~ type + sample
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsTypeSample <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c" 
                                                                                    &
                                                                                     matched == "Matched"),
                                       directory = directory,
                                       design = ~ type + sample)

ddsTypeSample <- estimateSizeFactors(ddsTypeSample)

# add annotation
rowData(ddsTypeSample)$ENS_gene_id <- rownames(rowData(ddsTypeSample))

normalizedCountsddsTypeSample <- counts(ddsTypeSample, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsTypeSample) >= 10) >= smallestGroupSize
ddsTypeSample <- ddsTypeSample[keep,]
# colSums(counts(ddsTypeSample)) %>% barplot
# ddsTypeSample$matched <- factor(ddsTypeSample$matched, levels = c("Matched", "Unmatched"))
ddsTypeSample$type <- factor(ddsTypeSample$type, levels = c("bulk_nuclei", "bulk_cell", "bulk_cell_p"))
```

## 1.3 Design ~ type + matched 
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsTypeMatched <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c" ),
                                       directory = directory,
                                       design = ~ type + matched)

ddsTypeMatched <- estimateSizeFactors(ddsTypeMatched)

# add annotation
rowData(ddsTypeMatched)$ENS_gene_id <- rownames(rowData(ddsTypeMatched))

normalizedCountsddsTypeMatched <- counts(ddsTypeMatched, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsTypeMatched) >= 10) >= smallestGroupSize
ddsTypeMatched <- ddsTypeMatched[keep,]
# colSums(counts(ddsTypeMatched)) %>% barplot

ddsTypeMatched$matched <- factor(ddsTypeMatched$matched, levels = c("Matched", "Unmatched"))
ddsTypeSample$type <- factor(ddsTypeSample$matched, levels = c("bulk_nuclei", "bulk_cell", "bulk_cell_p"))
```

## 1.3 Design ~  matched +typeOverall
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsTypeOverallMatched <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c"),
                                       directory = directory,
                                       design = ~ matched + typeOverall)

ddsTypeOverallMatched <- estimateSizeFactors(ddsTypeOverallMatched)

# add annotation
rowData(ddsTypeOverallMatched)$ENS_gene_id <- rownames(rowData(ddsTypeOverallMatched))

normalizedCountsddsTypeOverallMatched <- counts(ddsTypeOverallMatched, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsTypeOverallMatched) >= 10) >= smallestGroupSize
ddsTypeOverallMatched <- ddsTypeOverallMatched[keep,]
# colSums(counts(ddsTypeOverallMatched)) %>% barplot

ddsTypeOverallMatched$matched <- factor(ddsTypeOverallMatched$matched, levels = c("Matched", "Unmatched"))
```

## 1.3 Design ~ type + matched (without outlier)
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsTypeMatchedOutlier <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName %in% c("3145T1c", "3145T11a")),
                                       directory = directory,
                                       design = ~ type + matched)

ddsTypeMatchedOutlier <- estimateSizeFactors(ddsTypeMatchedOutlier)

# add annotation
rowData(ddsTypeMatchedOutlier)$ENS_gene_id <- rownames(rowData(ddsTypeMatchedOutlier))

normalizedCountsddsTypeMatchedOutlier <- counts(ddsTypeMatchedOutlier, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsTypeMatchedOutlier) >= 10) >= smallestGroupSize
ddsTypeMatchedOutlier <- ddsTypeMatchedOutlier[keep,]
# colSums(counts(ddsTypeMatched)) %>% barplot

ddsTypeMatchedOutlier$matched <- factor(ddsTypeMatchedOutlier$matched, levels = c("Matched", "Unmatched"))
```

## 1.4 Design ~  matched (bulk_cell + bulk_cell_p)
Bulk cell vs bulk nuc
```{r}
rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsMatched <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c" & typeOverall== "bulk_cell"),
                                       directory = directory,
                                       design = ~ type + matched)

ddsMatched <- estimateSizeFactors(ddsMatched)

# add annotation
rowData(ddsMatched)$ENS_gene_id <- rownames(rowData(ddsMatched))

normalizedCountsddsMatched <- counts(ddsMatched, normalized=TRUE)

smallestGroupSize <- 3
keep <- rowSums(counts(ddsMatched) >= 10) >= smallestGroupSize
ddsMatched <- ddsMatched[keep,]
# colSums(counts(ddsTypeMatched)) %>% barplot

# Control vs Treatment
ddsMatched$matched <- factor(ddsMatched$matched, levels = c("Unmatched", "Matched"))
# ddsMatched$typeOverall <- factor(ddsMatched$typeOverall, levels = c("bulk_nuc", "bulk_cell"))
```

# 3 Run DESEQ
## 3.1 Type (filter out matched)
```{r}
ddsType <- DESeq(ddsType)
resType <- results(ddsType)
resType

vsdataType <- vst(ddsType, blind=FALSE)
colData(ddsType)
plotPCA(vsdataType, intgroup=c("type"))

pdf(paste0(plotDir, "pca/", fprefix, "_type_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataType, intgroup=c("type")) + 
  geom_label(aes(label = name),size = 2)
dev.off()
```
## 3.2 Type Sample
```{r}
ddsTypeSample <- DESeq(ddsTypeSample)
resTypeSample <- results(ddsTypeSample)
resTypeSample

vsdataTypeSample <- vst(ddsTypeSample, blind=FALSE)
plotPCA(vsdataTypeSample, intgroup="type")

pdf(paste0(plotDir, "pca/", fprefix, "_typeSample_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataTypeSample, intgroup="type") + 
  geom_label(aes(label = name),size = 2)
dev.off()

```

## 3.3 Type Matched
```{r}
ddsTypeMatched <- DESeq(ddsTypeMatched)
resTypeMatched <- results(ddsTypeMatched)
resTypeMatched

vsdataTypeMatched <- vst(ddsTypeMatched, blind=FALSE)
plotPCA(vsdataTypeMatched, intgroup="matched") + geom_label(aes(label = name))

summary(resTypeMatched)

pdf(paste0(plotDir, "pca/", fprefix, "_typeMatched_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataTypeMatched, intgroup="matched") + 
  geom_label(aes(label = name),size = 2)
dev.off()

# out of 16020 with nonzero total read count
# adjusted p-value < 0.1
# LFC > 0 (up)       : 1366, 8.5%
# LFC < 0 (down)     : 1729, 11%
# outliers [1]       : 184, 1.1%
# low counts [2]     : 0, 0%
# (mean count < 2)
# [1] see 'cooksCutoff' argument of ?results
# [2] see 'independentFiltering' argument of ?results

```

## 3.4 Type Matched Outlier
```{r}
ddsTypeMatchedOutlier <- DESeq(ddsTypeMatchedOutlier)
resTypeMatchedOutlier <- results(ddsTypeMatchedOutlier)
resTypeMatchedOutlier

vsdataTypeMatchedOutlier <- vst(ddsTypeMatchedOutlier, blind=FALSE)

pdf(paste0(plotDir, "pca/", fprefix, "_typeMatchedOutlier_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataTypeMatchedOutlier, intgroup="matched") + 
  geom_label(aes(label = name),size = 2)
dev.off()


summary(resTypeMatchedOutlier)

# out of 16020 with nonzero total read count
# adjusted p-value < 0.1
# LFC > 0 (up)       : 1366, 8.5%
# LFC < 0 (down)     : 1729, 11%
# outliers [1]       : 184, 1.1%
# low counts [2]     : 0, 0%
# (mean count < 2)
# [1] see 'cooksCutoff' argument of ?results
# [2] see 'independentFiltering' argument of ?results

```

## 3.4 Matched
```{r}
ddsMatched <- DESeq(ddsMatched)
resMatched <- results(ddsMatched)
resMatched

vsdataMatched <- vst(ddsMatched, blind=FALSE)
plotPCA(vsdataMatched, intgroup="matched")

pdf(paste0(plotDir, "pca/", fprefix, "_matched_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataMatched, intgroup="matched") + 
  geom_label(aes(label = name),size = 2)
dev.off()
```

## 3.5 Type Overall (filter out matched)
```{r}
ddsTypeOverall <- DESeq(ddsTypeOverall)
resTypeOverall <- results(ddsTypeOverall)
resTypeOverall

vsdataTypeOverall <- vst(ddsTypeOverall, blind=FALSE)
plotPCA(vsdataTypeOverall, intgroup="matched")

pdf(paste0(plotDir, "pca/", fprefix, "_typeOverall_ntop500.pdf"),
    height=5, width=7)
plotPCA(vsdataTypeOverall, intgroup="typeOverall") + 
  geom_label(aes(label = name),size = 2)
dev.off()
```
## 3.6 Matched + typeOverall
```{r}
library(ggrepel)
ddsTypeOverallMatched <- DESeq(ddsTypeOverallMatched)
resTypeOverallMatched <- results(ddsTypeOverallMatched)
resTypeOverallMatched

vsdataTypeOverallMatched <- vst(ddsTypeOverallMatched, blind=FALSE)
plotPCA(vsdataTypeOverallMatched, intgroup="matched")

pdf(paste0(plotDir, "pca/", fprefix, "_typeOverallMatched_ntop500.pdf"),
    height=5, width=7)


plotPCA(vsdataTypeOverallMatched, intgroup = c("matched", "typeOverall")) +
  geom_label_repel(
    aes(
      label = name,
      fontface = ifelse(typeOverall == "bulk_cell", "italic", "plain")
    ),
    size = 2,
    box.padding = 0.3,  # space around labels
    point.padding = 0.2, # space around points
    max.overlaps = Inf   # show all labels, just repel
  )
dev.off()
```
# 2 Plot
## 2.1 Design ~ type
### 2.1.3 Plot count distribution
#### 2.1.3.1 Number of reads per sample
```{r}
plotDf <- data.frame(
  Sample = colnames(counts(ddsType)),
  TotalCounts = colSums(counts(ddsType)),
  Type = ddsType$type
) %>%
  mutate(Sample = factor(Sample, levels = unique(Sample)))

bySample <- ggplot(plotDf, aes(x = Sample, y = TotalCounts, fill = Type)) +
  geom_bar(stat = "identity") +
  # facet_grid(. ~ Sample, scales = "free_x", space = "free_x") +
  theme(strip.text = element_blank()) +
  # theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "DESEQ2 Read Counts",
       x = "Sample",
       y = "Read Count") +
  scale_fill_manual(
    values = c(
      # "polya" = "#66c2a5",
      "bulk_cell" = "#8da0cb",
      "bulk_cell_p" = "#8da0cb",
      "bulk_nuclei" = "maroon"
    ),
    name = "Library type"
  )

byLib <- ggplot(plotDf, aes(x = Sample, y = TotalCounts, fill = Type)) +
  geom_bar(stat = "identity") +
  # facet_grid(. ~ libraryType, scales = "free_x", space = "free_x") +
  theme(strip.text = element_blank()) +
  # theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "DESEQ2 Read Counts",
       x = "Sample",
       y = "Read Count") +
  scale_fill_manual(
    values = c(
      # "polya" = "#66c2a5",
      "bulk_cell" = "#8da0cb",
      "bulk_cell_p" = "#8da0cb",
      "bulk_nuclei" = "maroon"
    ),
    name = "Library type"
  )

pdf(paste0(plotDir, "/bar/", fprefix, "_stackedBar_readCounts.pdf"), height=5, width=10)
print(bySample)
print(byLib)
dev.off()

```

#### 2.1.3.1 Norm counts per sample
```{r}
plotDf <- data.frame(
  Sample = colnames(normalizedCountsDdsType),
  TotalCounts = colSums(normalizedCountsDdsType)
) %>%
  mutate(
  Type = ddsType$type
) %>%
  mutate(Sample = factor(Sample, levels = unique(Sample)))

bySample <- ggplot(plotDf, aes(x = Sample, y = TotalCounts, fill = Type)) +
  geom_bar(stat = "identity") +
  # facet_grid(. ~ Sample, scales = "free_x", space = "free_x") +
  theme(strip.text = element_blank()) +
  # theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "DESEQ2 Normalized Read Counts",
       x = "Sample",
       y = "Read Count") +
  scale_fill_manual(
    values = c(
      # "polya" = "#66c2a5",
      "bulk_cell" = "#8da0cb",
      "bulk_cell_p" = "#8da0cb",
      "bulk_nuclei" = "maroon"
    ),
    name = "Library type"
  )

byLib <- ggplot(plotDf, aes(x = Sample, y = TotalCounts, fill = Type)) +
  geom_bar(stat = "identity") +
  # facet_grid(. ~ libraryType, scales = "free_x", space = "free_x") +
  theme(strip.text = element_blank()) +
  # theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  labs(title = "DESEQ2 Normalized Read Counts",
       x = "Sample",
       y = "Read Count") +
  scale_fill_manual(
    values = c(
      # "polya" = "#66c2a5",
      "bulk_cell" = "#8da0cb",
      "bulk_cell_p" = "#8da0cb",
      "bulk_nuclei" = "maroon"
    ),
    name = "Library type"
  )


pdf(paste0(plotDir, "/bar/", fprefix, "_stackedBar_normCounts.pdf"), height=5, width=10)
print(bySample)
print(byLib)
dev.off()

```
# 4 Overlap with candidate marker genes
Ideally, marker genes will be applicable for matched and non -matched samples. 
Of note, however, the unmatched samples were picked as more immunologically interesting... is the increase in plasma cells biologically relevant or an artefact of 1 or 2 strongly expressed markers. The purity from genetic deconvolution should be able to give more information on this. 




Or do we filter DEGs with a p-value but not with LFC?
## 4.1 Marker gene lists
```{r}
masterMarkerGenes <- read.xlsx("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_masterMarkerGenes.xlsx")

geneDict <- masterMarkerGenes %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

uniqueGeneDict <- masterMarkerGenes %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

cibersortMin50Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr50_minGenes2to10/CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")

cibersortDictMin50 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin50Genes2to10$NAME) %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

cibersortUniqueDictMin50 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin50Genes2to10$NAME & Unique == "Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()


cibersortMin10Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr10_minGenes2to10/CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")

cibersortDictMin10 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin10Genes2to10$NAME) %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

cibersortUniqueDictMin10 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin10Genes2to10$NAME & Unique == "Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()
```

## 4.2 Overlap
### 4.2.1 Type (filter out non-matched)
```{r}
summary(resType)

resTypeExport <- resType %>%
  as.data.frame() %>%
    rownames_to_column("GeneID")

write.xlsx(resTypeExport,
  paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resType.xlsx"))

resType <- read.xlsx(paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resType.xlsx")) %>%
  column_to_rownames("GeneID")

resTypeDegs <- resType %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.5 & padj < 0.05) %>%
  rownames_to_column("GeneID") 

# intersect(resTypeDegs$GeneID, masterMarkerGenes$GeneID)
# > intersect(resTypeDegs$GeneID, masterMarkerGenes$GeneID)
# [1] "ENSCAFG00000032358" --> only one gene overlaps. 

# 426 upregulated genes in bulk cell vs bulk nuc. 24 are in the marker gene list (mostly caf/endo)
upType <- resType %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange >= 2.5 & padj < 0.05) %>%
  left_join(masterMarkerGenes) %>%
  drop_na("cluster")

# 1,021 downregulated genes in bulk cell vs bulk nuc. 3 genes in the marker gene list.

downType <- resType %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange <= -2.5 & padj < 0.05) %>%
    left_join(masterMarkerGenes)%>%
  drop_na("cluster")

curatedMarkers <- masterMarkerGenes %>%
  filter(GeneID != intersect(resTypeDegs$GeneID, masterMarkerGenes$GeneID))

curatedMarkersUnique <- masterMarkerGenes %>%
  filter(GeneID != intersect(resTypeDegs$GeneID, masterMarkerGenes$GeneID))
  
# intersect(resTypeDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# [1] "ENSCAFG00000032358"

curatedCib10Up <- cibersortDictMin10 %>%
  filter(GeneID == intersect(upType$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib10Down <- cibersortDictMin10 %>%
  filter(GeneID == intersect(downType$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

# curatedCib10Unique <- cibersortUniqueDictMin10

# intersect(resTypeDegs$GeneID, cibersortDictMin50$GeneID)

curatedCib50Up <- cibersortDictMin50 %>%
  filter(GeneID == intersect(upType$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib50Down <- cibersortDictMin50 %>%
  filter(GeneID == intersect(downType$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

# curatedCib50Unique <- cibersortUniqueDictMin50 %>%
#   filter(GeneID != intersect(resTypeDegs$GeneID, cibersortUniqueDictMin50$GeneID))
```

### 4.2.2 Type Overall
```{r}
summary(resTypeOverall)

resTypeOverallExport <- resTypeOverall %>%
  as.data.frame() %>% 
    rownames_to_column("GeneID")

write.xlsx(resTypeOverallExport, 
  paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resTypeOverall.xlsx"))

resTypeOverallDegs <- resTypeOverall %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
  rownames_to_column("GeneID") 

# intersect(resTypeOverallDegs$GeneID, masterMarkerGenes$GeneID)
# > intersect(resTypeOverallDegs$GeneID, masterMarkerGenes$GeneID)
# [1] "ENSCAFG00000032358" --> only one gene overlaps. 

# 626 upregulated genes in cell vs nuc. 44 in the marker gene list. 
upTypeOverall <- resTypeOverall %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange >= 2.5 & padj < 0.05) %>%
  left_join(masterMarkerGenes) %>%
  drop_na("cluster")

# 23 downregulated genes in cell vs nuc. None are in the marker gene list

downTypeOverall <- resTypeOverall %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange <= -2.5 & padj < 0.05) %>%
    left_join(masterMarkerGenes)%>%
  drop_na("cluster")

curatedMarkers <- masterMarkerGenes %>%
  filter(GeneID != intersect(resTypeOverallDegs$GeneID, masterMarkerGenes$GeneID))

curatedMarkersUnique <- masterMarkerGenes %>%
  filter(GeneID != intersect(resTypeOverallDegs$GeneID, masterMarkerGenes$GeneID))
  
# intersect(resTypeOverallDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# [1] "ENSCAFG00000032358"

curatedCib10Up <- cibersortDictMin10 %>%
  filter(GeneID == intersect(upTypeOverall$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib10Down <- cibersortDictMin10 %>%
  filter(GeneID == intersect(downTypeOverall$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

# curatedCib10Unique <- cibersortUniqueDictMin10

# intersect(resTypeDegs$GeneID, cibersortDictMin50$GeneID)

curatedCib50Up <- cibersortDictMin50 %>%
  filter(GeneID == intersect(upTypeOverall$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib50Down <- cibersortDictMin50 %>%
  filter(GeneID == intersect(downTypeOverall$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)
```

### 4.2.2 Matched
```{r}
summary(resMatched)

resMatchedExport <- resMatched %>%
  as.data.frame() %>%
    rownames_to_column("GeneID")

# write.xlsx(resMatchedExport, 
#   paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resMatched.xlsx"))

resMatched <- read.xlsx(paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resMatched.xlsx")) %>%
  column_to_rownames("GeneID")

resMatchedDegs <- resMatched %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.5 & padj < 0.05) %>%
  rownames_to_column("GeneID") 

# intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# > intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# [1] "ENSCAFG00000032358" --> only one gene overlaps. 

# 106 upregulated genes in Matched vs Unmatched. None are in the marker gene list. 
upMatched <- resMatched %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange >= 2.5 & padj < 0.05) %>%
  left_join(masterMarkerGenes) %>%
  drop_na("cluster")

# 89 downregulated genes in Matched vs Unmatched. One gene is in the marker gene list: Plasma DEG ENSCAFG00000032358

downMatched <- resMatched %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange <= -2.5 & padj < 0.05) %>%
    left_join(masterMarkerGenes)%>%
  drop_na("cluster")

curatedMarkers <- masterMarkerGenes %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID))

curatedMarkersUnique <- masterMarkerGenes %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID))
  
# intersect(resMatchedDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# [1] "ENSCAFG00000032358"

curatedCib10Up <- cibersortDictMin10 %>%
  filter(GeneID == intersect(upMatched$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib10Down <- cibersortDictMin10 %>%
  filter(GeneID == intersect(downMatched$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

# curatedCib10Unique <- cibersortUniqueDictMin10

# intersect(resTypeDegs$GeneID, cibersortDictMin50$GeneID)

curatedCib50Up <- cibersortDictMin50 %>%
  filter(GeneID == intersect(upMatched$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)

curatedCib50Down <- cibersortDictMin50 %>%
  filter(GeneID == intersect(downMatched$GeneID, cibersortDictMin50$GeneID)) %>%
  left_join(masterMarkerGenes)
```
### 4.2.3 Type Matched
```{r}
summary(resMatched)

resMatchedExport <- resMatched %>%
  as.data.frame() %>% 
    rownames_to_column("GeneID")

write.xlsx(resMatchedExport, 
  paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resMatched.xlsx"))

resMatchedDegs <- resMatched %>%
  as.data.frame() %>%
  filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
  rownames_to_column("GeneID") 

# intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# > intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# [1] "ENSCAFG00000032358" --> only one gene overlaps. 

# 95 genes --> Only one gene in the sig. matrix table. ENSCAFG00000032358
upMatched <- resMatched %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange >= 2 & padj < 0.05) %>%
  left_join(masterMarkerGenes) %>%
  drop_na("cluster")

# 117 --> None are in the sig. matrix table. 
downMatched <- resMatched %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(log2FoldChange <= -2 & padj < 0.05) %>%
    left_join(masterMarkerGenes)%>%
  drop_na("cluster")

curatedMarkers <- masterMarkerGenes %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID))

curatedMarkersUnique <- masterMarkerGenes %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, masterMarkerGenes$GeneID))
  
# intersect(resMatchedDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# [1] "ENSCAFG00000032358"

curatedCib10 <- cibersortDictMin10 

curatedCib10Unique <- cibersortUniqueDictMin10

intersect(resMatchedDegs$GeneID, cibersortDictMin50$GeneID)

curatedCib50 <- cibersortDictMin50 %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, cibersortDictMin50$GeneID))

curatedCib50Unique <- cibersortUniqueDictMin50 %>%
  filter(GeneID != intersect(resMatchedDegs$GeneID, cibersortUniqueDictMin50$GeneID))
```

### 4.2.4 Type Matched Outlier
```{r}
# summary(resMatchedOutlier)
# 
# write.xlsx(resMatchedOutlier%>%
#   as.data.frame(), paste0(resDir, fprefix, "_resMatchedOutlier.xlsx"))
# 
# resMatchedOutlierDegs <- resMatchedOutlier %>%
#   as.data.frame() %>%
#   filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
#   rownames_to_column("GeneID") 
# 
# # intersect(resMatchedOutlierDegs$GeneID, masterMarkerGenes$GeneID)
# # > intersect(resMatchedOutlierDegs$GeneID, masterMarkerGenes$GeneID)
# # [1] "ENSCAFG00000032358" --> only one gene overlaps. 
# 
# # 95 genes --> Only one gene in the sig. matrix table. 
# upMatched <- resMatchedOutlier %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(log2FoldChange >= 2 & padj < 0.05) %>%
#   left_join(masterMarkerGenes)
# 
# # 117 --> None are in the sig. matrix table. 
# downMatched <- resMatchedOutlier %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(log2FoldChange <= -2 & padj < 0.05) %>%
#     left_join(masterMarkerGenes)
# 
# curatedMarkers <- masterMarkerGenes %>%
#   filter(GeneID != intersect(resMatchedOutlierDegs$GeneID, masterMarkerGenes$GeneID))
# 
# curatedMarkersUnique <- masterMarkerGenes %>%
#   filter(GeneID != intersect(resMatchedOutlierDegs$GeneID, masterMarkerGenes$GeneID))
#   
# # intersect(resMatchedOutlierDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# # [1] "ENSCAFG00000032358"
# 
# curatedCib10 <- cibersortDictMin10 
# 
# curatedCib10Unique <- cibersortUniqueDictMin10
# 
# intersect(resMatchedOutlierDegs$GeneID, cibersortDictMin50$GeneID)
# 
# curatedCib50 <- cibersortDictMin50 %>%
#   filter(GeneID != intersect(resMatchedOutlierDegs$GeneID, cibersortDictMin50$GeneID))
# 
# curatedCib50Unique <- cibersortUniqueDictMin50 %>%
#   filter(GeneID != intersect(resMatchedOutlierDegs$GeneID, cibersortUniqueDictMin50$GeneID))
```

### 4.2.5 typeOverall Matched
```{r}
# summary(resTypeOverallMatched)
# 
# resTypeOverallMatchedExport <- resTypeOverallMatched %>%
#   as.data.frame() %>% 
#     rownames_to_column("GeneID")
# 
# write.xlsx(resTypeOverallMatchedExport, 
#   paste0( "~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_DESEQ/03_Deconvolution_CTVT_DESEQ_resTypeOverallMatched.xlsx"))
# 
# resTypeOverallMatchedDegs <- resTypeOverallMatched %>%
#   as.data.frame() %>%
#   filter(abs(log2FoldChange) >= 2.0 & padj < 0.05) %>%
#   rownames_to_column("GeneID") 
# 
# # intersect(resTypeOverallMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# # > intersect(resTypeOverallMatchedDegs$GeneID, masterMarkerGenes$GeneID)
# # [1] "ENSCAFG00000032358" --> only one gene overlaps. 
# 
# # 61 DEGS --> None are in the sig. matrix table. 
# upMatched <- resTypeOverallMatched %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(log2FoldChange >= 2 & padj < 0.05) %>%
#   left_join(masterMarkerGenes) %>%
#   drop_na("cluster")
# 
# # 117 --> None are in the sig. matrix table. 
# # 753 DEGS Down in , 45 
# downMatched <- resTypeOverallMatched %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(log2FoldChange <= -2 & padj < 0.05) %>%
#     left_join(masterMarkerGenes)%>%
#   drop_na("cluster")
# 
# # 45 DEGs
# allMatched <- resTypeOverallMatched %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(abs(log2FoldChange) >= 2 & padj < 0.05) %>%
#     left_join(masterMarkerGenes)%>%
#   drop_na("cluster")
# 
# curatedMarkers <- masterMarkerGenes %>%
#   filter(GeneID != intersect(resTypeOverallMatchedDegs$GeneID, masterMarkerGenes$GeneID))
# 
# curatedMarkersUnique <- masterMarkerGenes %>%
#   filter(GeneID != intersect(resTypeOverallMatchedDegs$GeneID, masterMarkerGenes$GeneID))
#   
# # intersect(resTypeOverallMatchedDegs$GeneID, cibersortDictMin10$GeneID) --> 0
# # [1] "ENSCAFG00000032358"
# 
# curatedCib10 <- cibersortDictMin10 
# 
# curatedCib10Unique <- cibersortUniqueDictMin10
# 
# intersect(resTypeOverallMatchedDegs$GeneID, cibersortDictMin50$GeneID)
# 
# curatedCib50 <- cibersortDictMin50 %>%
#   filter(GeneID != intersect(resTypeOverallMatchedDegs$GeneID, cibersortDictMin50$GeneID))
# 
# curatedCib50Unique <- cibersortUniqueDictMin50 %>%
#   filter(GeneID != intersect(resTypeOverallMatchedDegs$GeneID, cibersortUniqueDictMin50$GeneID))
```

# 5 Visualizations
```{r}
obj <- readRDS(paste0(dataDir, "SCE_concat_CTVT_includingAllGenes_logCorrecDone_vstCorrecDone_ClustDone2_ManAnot.rds"))

objNoBad <- obj[,obj$manual.log.anot.fine!="bad"]

rm(obj)
gc()

rowData <- rowData(objNoBad) %>%
  as.data.frame()

rm(objNoBad)
gc()

rownames(sampleTableI) <- sampleTableI$fileName
file.exists(file.path(directory, sampleTableI$fileName)) # check all files exists (here 3145T1c not yet here)

#create dds object (gather htseq count output)
ddsHT_SC <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTableI %>% filter(!sampleName == "3145T1c"),
                                       directory = directory,
                                       design = ~ type)

ddsHT_SC <- estimateSizeFactors(ddsHT_SC)

# add annotation
rowData(ddsHT_SC)$ENS_gene_id <- rownames(rowData(ddsHT_SC))
```

## 5.1 Heatmaps 
### 5.1.1 Marker genes
#### 5.1.1.1 Gene list
```{r}
masterMarkerGenes <- read.xlsx("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_masterMarkerGenes.xlsx")

cibersortDict <- masterMarkerGenes %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID, cluster) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct() 

B_cell <- cibersortDict %>%
  filter(cluster == "Bcells") %>%
  select(Symbol) %>%
  unlist()

T_cell <- cibersortDict %>%
  filter(cluster == "Tcells") %>%
  select(Symbol) %>%
  unlist()

Plasma <- cibersortDict %>%
  filter(cluster == "Plasma") %>%
  select(Symbol) %>%
  unlist()

CAF_Endo <- cibersortDict %>%
  filter(cluster == "CAF.Endo") %>%
  select(Symbol) %>%
  unlist()

CTVT <- cibersortDict %>%
  filter(cluster == "CTVT") %>%
  select(Symbol) %>%
  unlist()

Myeloid <- cibersortDict %>%
  filter(cluster == "Myeloid") %>%
  select(Symbol) %>%
  unlist()

```
#### 5.1.1.2 Plot
```{r, echo = F, message = FALSE, warning= FALSE}

immune_set <- list(
  `B_cell` = B_cell,
  `T_cell` = T_cell,
  # `MHCI` = c("TAP1", "B2M", "DLA-88", "DLA-64"),
  # `MHCII`= c("DLA-DRA", "DLA-DQA1", "DLA-DMA"),
  `Plasma` = Plasma,
  `Myeloid` = Myeloid,
  `CAF.Endo`= CAF_Endo,
  `CTVT` = CTVT 
)


row_cols <- list(
  cell_type =c(
  "CTVT" = "#BCD0C7",
  "CAF.Endo" = "#FDDFA4FF",
  "Myeloid" = "#F5FDC6",
  "Plasma" = "#B4B9E0FF",
  "T_cell" = "#D56AA0",
  "B_cell" = "#A8D5E2"),
  cell = c(cancer = "black",
           host = "white"),
  Origin = c(Nepal = "grey10",
             UK = "grey80",
             USA = "grey30"),
  type = c(bulk_cell = "black",
           bulk_cell_p = "grey40", 
           bulk_nuclei = "grey"),
  CTVT_clade = c(CTVTA = "darkred",
                 CTVTBG = "darkorange",
                 CTVTABG = "darkorange4",
                 tbc = "white",
                 CTVTE = "darkgreen",
                 CTVTD = "cyan4",
                 CTVTG = "purple4")
)

# Convert the list to a dataframe
immune_df <- stack(immune_set) %>% 
  as.data.frame() %>%
  dplyr::rename(geneID = values, cell_type = ind) %>%
  mutate(cell = case_when(
    cell_type == 'CTVT' ~ "cancer",
    TRUE  ~ "host"
  )) %>%
  mutate(cell = factor(cell, levels = c("cancer", 
                                        "host"))) %>%
  #mutate(cell_type = factor(cell_type, levels = names(row_cols$cell_type))) %>%
  mutate(geneID = factor(geneID))
  
rownames(immune_df) <- immune_df$geneID

# Rename bulk data. 


# heatmap on vst
vsd <- vst(ddsHT_SC, blind=T, fitType = "mean")

mat <- assay(vsd) %>%
  as.data.frame() %>%
  rownames_to_column("ID") %>%
  filter(ID %in% rowData$ID) %>%
  left_join(rowData) %>%
  filter(Symbol %in% rownames(immune_df)) %>%
  dplyr::select(-c(ID, Type, Chromosome, hvgs)) %>%
  column_to_rownames("Symbol") %>%
  as.matrix()

de_mat <- mat[rownames(mat) %in% immune_df$geneID,]

de_mat <- de_mat[match(rownames(immune_df[immune_df$geneID %in% rownames(mat),]), rownames(de_mat)),]



a <- pheatmap::pheatmap(t(scale(t(de_mat))),
                   show_rownames = T,
                   show_colnames = T,
                   cluster_rows = F,
                   cluster_cols = F,
                   fontsize_row = 5 , 
                   gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
                   annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
                   annotation_row = immune_df[immune_df$geneID %in% rownames(mat), c("cell_type", "cell")],
                   annotation_colors = row_cols,
                   main = paste0("Cib10 Marker Genes"))

# b <- pheatmap::pheatmap(de_mat, 
#                    show_rownames = T,
#                    show_colnames = T,
#                    cluster_rows = F,
#                    cluster_cols = F,
#                    fontsize_row = 5, 
#                    gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
#                    annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
#                    annotation_row = immune_df[immune_df$geneID %in% rownames(ddsHT_SC), c("cell_type", "cell") ],
#                    annotation_colors = row_cols,
#                    main = paste0("Heatmap of immune cell related genes, NOT scaled by gene"))


pdf(paste0(plotDir, "heatmap/", fprefix, "_bulkExpression_markerGenesUnique_vst.pdf"), 
    height=10, width = 8)
print(a)
# print(b)
dev.off()


```

### 5.1.2 Cib10 
#### 5.1.3.1 Gene list
```{r}
cibersortMin10Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr10_minGenes2to10/CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job42_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")


cibersortDictMin10 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin10Genes2to10$NAME) %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID, cluster) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct() 

B_cell <- cibersortDictMin10 %>%
  filter(cluster == "Bcells") %>%
  select(Symbol) %>%
  unlist()

T_cell <- cibersortDictMin10 %>%
  filter(cluster == "Tcells") %>%
  select(Symbol) %>%
  unlist()

Plasma <- cibersortDictMin10 %>%
  filter(cluster == "Plasma") %>%
  select(Symbol) %>%
  unlist()

CAF_Endo <- cibersortDictMin10 %>%
  filter(cluster == "CAF.Endo") %>%
  select(Symbol) %>%
  unlist()

CTVT <- cibersortDictMin10 %>%
  filter(cluster == "CTVT") %>%
  select(Symbol) %>%
  unlist()

Myeloid <- cibersortDictMin10 %>%
  filter(cluster == "Myeloid") %>%
  select(Symbol) %>%
  unlist()
```
#### 5.1.3.2 Plot
```{r, echo = F, message = FALSE, warning= FALSE}

immune_set <- list(
  `B_cell` = B_cell,
  `T_cell` = T_cell,
  # `MHCI` = c("TAP1", "B2M", "DLA-88", "DLA-64"),
  # `MHCII`= c("DLA-DRA", "DLA-DQA1", "DLA-DMA"),
  `Plasma` = Plasma,
  `Myeloid` = Myeloid,
  `CAF.Endo`= CAF_Endo,
  `CTVT` = CTVT 
)


row_cols <- list(
  cell_type =c(
  "CTVT" = "#BCD0C7",
  "CAF.Endo" = "#FDDFA4FF",
  "Myeloid" = "#F5FDC6",
  "Plasma" = "#B4B9E0FF",
  "T_cell" = "#D56AA0",
  "B_cell" = "#A8D5E2"),
  cell = c(cancer = "black",
           host = "white"),
  Origin = c(Nepal = "grey10",
             UK = "grey80",
             USA = "grey30"),
  type = c(bulk_cell = "black",
           bulk_cell_p = "grey40", 
           bulk_nuclei = "grey"),
  CTVT_clade = c(CTVTA = "darkred",
                 CTVTBG = "darkorange",
                 CTVTABG = "darkorange4",
                 tbc = "white",
                 CTVTE = "darkgreen",
                 CTVTD = "cyan4",
                 CTVTG = "purple4")
)

# Convert the list to a dataframe
immune_df <- stack(immune_set) %>% 
  as.data.frame() %>%
  dplyr::rename(geneID = values, cell_type = ind) %>%
  mutate(cell = case_when(
    cell_type == 'CTVT' ~ "cancer",
    TRUE  ~ "host"
  )) %>%
  mutate(cell = factor(cell, levels = c("cancer", 
                                        "host"))) %>%
  #mutate(cell_type = factor(cell_type, levels = names(row_cols$cell_type))) %>%
  mutate(geneID = factor(geneID))
  
rownames(immune_df) <- immune_df$geneID

# Rename bulk data. 


# heatmap on vst
vsd <- vst(ddsHT_SC, blind=T, fitType = "mean")

mat <- assay(vsd) %>%
  as.data.frame() %>%
  rownames_to_column("ID") %>%
  filter(ID %in% rowData$ID) %>%
  left_join(rowData) %>%
  filter(Symbol %in% rownames(immune_df)) %>%
  dplyr::select(-c(ID, Type, Chromosome, hvgs)) %>%
  column_to_rownames("Symbol") %>%
  as.matrix()

de_mat <- mat[rownames(mat) %in% immune_df$geneID,]

de_mat <- de_mat[match(rownames(immune_df[immune_df$geneID %in% rownames(mat),]), rownames(de_mat)),]



a <- pheatmap::pheatmap(t(scale(t(de_mat))),
                   show_rownames = T,
                   show_colnames = T,
                   cluster_rows = F,
                   cluster_cols = F,
                   fontsize_row = 5 , 
                   gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
                   annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
                   annotation_row = immune_df[immune_df$geneID %in% rownames(mat), c("cell_type", "cell")],
                   annotation_colors = row_cols,
                   main = paste0("Cib10 Marker Genes"))

# b <- pheatmap::pheatmap(de_mat, 
#                    show_rownames = T,
#                    show_colnames = T,
#                    cluster_rows = F,
#                    cluster_cols = F,
#                    fontsize_row = 5, 
#                    gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
#                    annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
#                    annotation_row = immune_df[immune_df$geneID %in% rownames(ddsHT_SC), c("cell_type", "cell") ],
#                    annotation_colors = row_cols,
#                    main = paste0("Heatmap of immune cell related genes, NOT scaled by gene"))


pdf(paste0(plotDir, "heatmap/", fprefix, "_bulkExpression_cib10Unique_vst.pdf"), 
    height=10, width = 8)
print(a)
# print(b)
dev.off()


```

### 5.1.3 Cib50 (unique)
#### 5.1.3.1 Gene list
```{r}
cibersortMin50Genes2to10 <- read_table("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_cibersort/snCountsFinalq1q4/sigMatrix_minExpr50_minGenes2to10/CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt")

cibersortDictMin50 <- masterMarkerGenes %>%
  filter(Symbol %in% cibersortMin50Genes2to10$NAME) %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID, cluster) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct() 

B_cell <- cibersortDictMin50 %>%
  filter(cluster == "Bcells") %>%
  select(Symbol) %>%
  unlist()

T_cell <- cibersortDictMin50 %>%
  filter(cluster == "Tcells") %>%
  select(Symbol) %>%
  unlist()

Plasma <- cibersortDictMin50 %>%
  filter(cluster == "Plasma") %>%
  select(Symbol) %>%
  unlist()

CAF_Endo <- cibersortDictMin50 %>%
  filter(cluster == "CAF.Endo") %>%
  select(Symbol) %>%
  unlist()

CTVT <- cibersortDictMin50 %>%
  filter(cluster == "CTVT") %>%
  select(Symbol) %>%
  unlist()

Myeloid <- cibersortDictMin50 %>%
  filter(cluster == "Myeloid") %>%
  select(Symbol) %>%
  unlist()
```

#### 5.1.3.3 Plot
```{r, echo = F, message = FALSE, warning= FALSE}

immune_set <- list(
  `B_cell` = B_cell,
  `T_cell` = T_cell,
  # `MHCI` = c("TAP1", "B2M", "DLA-88", "DLA-64"),
  # `MHCII`= c("DLA-DRA", "DLA-DQA1", "DLA-DMA"),
  `Plasma` = Plasma,
  `Myeloid` = Myeloid,
  `CAF.Endo`= CAF_Endo,
  `CTVT` = CTVT 
)


row_cols <- list(
  cell_type =c(
  "CTVT" = "#BCD0C7",
  "CAF.Endo" = "#FDDFA4FF",
  "Myeloid" = "#F5FDC6",
  "Plasma" = "#B4B9E0FF",
  "T_cell" = "#D56AA0",
  "B_cell" = "#A8D5E2"),
  cell = c(cancer = "black",
           host = "white"),
  Origin = c(Nepal = "grey50",
             UK = "grey80",
             USA = "grey30"),
  type = c(bulk_cell = "black",
           bulk_cell_p = "grey40", 
           bulk_nuclei = "grey"),
  CTVT_clade = c(CTVTA = "darkred",
                 CTVTBG = "darkorange",
                 CTVTABG = "darkorange4",
                 tbc = "white",
                 CTVTE = "darkgreen",
                 CTVTD = "cyan4",
                 CTVTG = "purple4")
)

# Convert the list to a dataframe
immune_df <- stack(immune_set) %>% 
  as.data.frame() %>%
  dplyr::rename(geneID = values, cell_type = ind) %>%
  mutate(cell = case_when(
    cell_type == 'CTVT' ~ "cancer",
    TRUE  ~ "host"
  )) %>%
  mutate(cell = factor(cell, levels = c("cancer", 
                                        "host"))) %>%
  #mutate(cell_type = factor(cell_type, levels = names(row_cols$cell_type))) %>%
  mutate(geneID = factor(geneID))
  
rownames(immune_df) <- immune_df$geneID

# Rename bulk data. 


# heatmap on vst
vsd <- vst(ddsHT_SC, blind=T, fitType = "mean")

mat <- assay(vsd) %>%
  as.data.frame() %>%
  rownames_to_column("ID") %>%
  filter(ID %in% rowData$ID) %>%
  left_join(rowData) %>%
  filter(Symbol %in% rownames(immune_df)) %>%
  dplyr::select(-c(ID, Type, Chromosome, hvgs)) %>%
  column_to_rownames("Symbol") %>%
  as.matrix()

de_mat <- mat[rownames(mat) %in% immune_df$geneID,]

de_mat <- de_mat[match(rownames(immune_df[immune_df$geneID %in% rownames(mat),]), rownames(de_mat)),]



a <- pheatmap::pheatmap(t(scale(t(de_mat))),
                   show_rownames = T,
                   show_colnames = T,
                   cluster_rows = F,
                   cluster_cols = F,
                   fontsize_row = 5 , 
                   gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
                   annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
                   annotation_row = immune_df[immune_df$geneID %in% rownames(mat), c("cell_type", "cell")],
                   annotation_colors = row_cols,
                   main = paste0("Cib50 Marker Genes"))

# b <- pheatmap::pheatmap(de_mat, 
#                    show_rownames = T,
#                    show_colnames = T,
#                    cluster_rows = F,
#                    cluster_cols = F,
#                    fontsize_row = 5, 
#                    gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(mat), ]) %>% as.vector()),
#                    annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
#                    annotation_row = immune_df[immune_df$geneID %in% rownames(ddsHT_SC), c("cell_type", "cell") ],
#                    annotation_colors = row_cols,
#                    main = paste0("Heatmap of immune cell related genes, NOT scaled by gene"))


pdf(paste0(plotDir, "heatmap/", fprefix, "_bulkExpression_cib50Unique_vst.pdf"), 
    height=10, width = 8)
print(a)
# print(b)
dev.off()


```



**PCA LOG normalisation**

PCA on normalised counts using **log(counts(dds, normalized = T) + 1)**. The number of PC axis created correspond to the number of samples. All the genes are selected for the PCA (also tried top 500 most variable genes, and give similar results). I use a modified plotPCA function in order to keep all info about samples and genes contribution for all axis, and plot manually all axis needed.

```{r, echo = F, message = FALSE, warning= FALSE}


# PCA "homemade", LOG
pcaData <- plotPCA.DEseq3(ddsHT_SC,
                             intgroup=c("sample", "type"),
                             method_norm = "log",
                             returnData = T,
                             ntop = nrow(ddsHT_SC),
                             scale = F) # function modif to access all info and axis wanted
percentVar <- round(100 * attr(pcaData, "percentVar"))

geneContrib <- attr(pcaData, "GeneContrib")

# PC 1-2
p1 <- ggplot(pcaData,
             aes(color = sample,
                 shape = type,
                 group = type,
                 x = PC1,
                 y = PC2)) +
  geom_hline(yintercept = 0, linetype = "longdash", color = "grey") +
  geom_vline(xintercept = 0, linetype = "longdash", color = "grey") +
  geom_point(size = 3, alpha = 0.9) +
  xlab(paste0("PCA1 (", percentVar[1],"%)")) +
  ylab(paste0("PCA2 (", percentVar[2],"%)")) +
  ggtitle("PCA axis 1 and 2 - LOG normalisation") +
  guides(fill = "none", color = "none", shape ="none") +
  cowplot::theme_cowplot()

plotly::ggplotly(p1)


# all 1:5
p <- GGally::ggpairs(pcaData,
              columns = 1:5, # vérifier l'organisation du fichier pcaData, ici programmer pour 10 axes PCA
              aes(color = sample,
                  shape = type,
                  group = type
                  ),
              title = "PCA axis 1 to 5 - LOG normalisation",
              diag = list(continuous = "densityDiag"),
              #lower = "blank",
              upper = list(continuous = "points")) +
  scale_color_manual(values = col_Ind) +
  theme_bw()
plotly::ggplotly(p)

# all 6:10
p <- GGally::ggpairs(pcaData,
              columns = 6:10, # vérifier l'organisation du fichier pcaData, ici programmer pour 10 axes PCA
              aes(color = sample,
                  shape = type,
                  group = type
                  ),
              title = "PCA axis 6 to 10 - VST normalisation",
              diag = list(continuous = "densityDiag"),
              #lower = "blank",
              upper = list(continuous = "points")) +
  scale_color_manual(values = col_Ind) +
  theme_bw()
plotly::ggplotly(p)

# 3D 1,2,3
plotly::plot_ly(data = pcaData,
                x = ~PC1,
                y = ~PC2,
                z = ~PC3,
                type = "scatter3d",
                color = ~sample,
                symbol = ~type,
                symbols = c("circle", "circle-open"), #one of ( "circle" | "circle-open" | "cross" | "diamond" | "diamond-open" | "square" | "square-open" | "x" )
                text = ~sample + type,
                colors = col_Ind) 


```


.
Let's check the heatmaps with log normalisation. 
.


```{r, echo = F, message = FALSE, warning= FALSE}
geneContrib2 <- geneContrib %>% 
  as_tibble() %>%
  mutate(GeneID = rownames(geneContrib)) %>%
  dplyr::select(GeneID, everything())

PCheatmap <- function(dds, n=n, PC = PC){
  
  ann_cols <- list(
  Origin = c(Nepal = "grey50",
             UK = "grey80",
             USA = "grey30"),
  type = c(bulk_cell = "black",
           bulk_cell_p = "grey40", 
           bulk_nuclei = "grey"),
  CTVT_clade = c(CTVTA = "darkred",
                 CTVTBG = "darkorange",
                 CTVTABG = "darkorange4",
                 tbc = "white",
                 CTVTE = "darkgreen",
                 CTVTD = "cyan4",
                 CTVTG = "purple4"))
  
PC <- PC
PC_pos <- geneContrib2 %>% slice_max(.[,PC] ,n = n) %>% pull(GeneID)
PC_neg <- geneContrib2 %>% slice_min(.[,PC] ,n = n) %>% pull(GeneID)

de_mat <- log(counts(ddsHT_SC, normalize = T)[rownames(ddsHT_SC) %in% c(PC_neg, PC_pos),] + 1)

pheatmap::pheatmap(t(scale(t(de_mat))),
                   show_rownames = T,
                   show_colnames = T,
                   fontsize_row = 5, 
                   annotation_col = as.data.frame(colData(dds)[,c("type","CTVT_clade", "Origin")]),
                   annotation_colors = ann_cols,
                   main = paste0("Heatmap of gene contribution to ", PC, " (top", n, " positive/negative), scaled by gene"))

}

PCheatmap(ddsHT_SC, n = 20, PC = "PC1")
PCheatmap(ddsHT_SC, n = 20, PC = "PC2")
PCheatmap(ddsHT_SC, n = 20, PC = "PC3")
PCheatmap(ddsHT_SC, n = 20, PC = "PC4")
PCheatmap(ddsHT_SC, n = 20, PC = "PC5")
PCheatmap(ddsHT_SC, n = 20, PC = "PC6")
PCheatmap(ddsHT_SC, n = 20, PC = "PC7")
PCheatmap(ddsHT_SC, n = 20, PC = "PC8")
PCheatmap(ddsHT_SC, n = 20, PC = "PC9")
PCheatmap(ddsHT_SC, n = 20, PC = "PC10")



```


.
Now let's check marker genes of immune cells types found in snRNAseq or known in litterature (vst, scaled by genes or not)
.


```{r, echo = F, message = FALSE, warning= FALSE}

immune_set <- list(
  `B_cell` = c("MS4A1", "PAX5", "BANK1", "BCL11A"),
  `T_cell` = c("CD3E", "CD3G", "CD4", "CD8A", "ICOS", "CTLA4"),
  `MHCI` = c("TAP1", "B2M", "DLA-88", "DLA-64"),
  `MHCII`= c("DLA-DRA", "DLA-DQA1", "DLA-DMA"),
  `Plasma` = c("JCHAIN", "IRF4", "POU2AF1"),
  `Myeloid` = c("LYZ", "CD80", "CD86", "IL18", "MRC1", "GAB2"),
  `CAFs`= c("FAP", "CALD1", "COL6A3", "LAMA4"),
  `Endo` = c("ADGRL4", "VWF", "PDGFD"),
  `CTVT` = c("LSAMP", "DPP10", "PDE1C", "PDGFC", "DLC1", "EPO", "MS4A2", "KIT") 
  
)

row_cols <- list(
  cell_type =c( T_cell = "skyblue1",#fine
                B_cell = "slategray3",#fine
                MHCI = "pink",
                MHCII = "pink3",
                Plasma = "lightblue1",#fine
                Myeloid = "mediumpurple3",
                CAFs = "darkolivegreen4",
                Endo = "tan3",
                CTVT = "firebrick4"),
  cell = c(cancer = "black",
           host = "white"),
  Origin = c(Nepal = "grey50",
             UK = "grey80",
             USA = "grey30"),
  type = c(bulk_cell = "black",
           bulk_cell_p = "grey40", 
           bulk_nuclei = "grey"),
  CTVT_clade = c(CTVTA = "darkred",
                 CTVTBG = "darkorange",
                 CTVTABG = "darkorange4",
                 tbc = "white",
                 CTVTE = "darkgreen",
                 CTVTD = "cyan4",
                 CTVTG = "purple4")
)

# Convert the list to a dataframe
immune_df <- stack(immune_set) %>% 
  as.data.frame() %>%
  dplyr::rename(geneID = values, cell_type = ind) %>%
  mutate(cell = case_when(
    cell_type == 'CTVT' ~ "cancer",
    TRUE  ~ "host"
  )) %>%
  mutate(cell = factor(cell, levels = c("cancer", "host"))) %>%
  #mutate(cell_type = factor(cell_type, levels = names(row_cols$cell_type))) %>%
  mutate(geneID = factor(geneID))
  
rownames(immune_df) <- immune_df$geneID


# heatmap on log(counts + 1)
de_mat <- log(counts(ddsHT_SC, normalize = T)[rownames(ddsHT_SC) %in% immune_df$geneID,] + 1)
de_mat <- de_mat[match(rownames(immune_df[immune_df$geneID %in% rownames(ddsHT_SC),]), rownames(de_mat)),]
pheatmap::pheatmap(t(scale(t(de_mat))),
                   show_rownames = T,
                   show_colnames = T,
                   cluster_rows = F,
                   cluster_cols = F,
                   fontsize_row = 5, 
                   gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(ddsHT_SC), ]) %>% as.vector()),
                   annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
                   annotation_row = immune_df[immune_df$geneID %in% rownames(ddsHT_SC), c("cell_type", "cell")],
                   annotation_colors = row_cols,
                   main = paste0("Heatmap of immune cell related genes, scaled by gene"))

pheatmap::pheatmap(de_mat, 
                   show_rownames = T,
                   show_colnames = T,
                   cluster_rows = F,
                   cluster_cols = F,
                   fontsize_row = 5, 
                   gaps_row = cumsum(xtabs(~cell_type, immune_df[immune_df$geneID %in% rownames(ddsHT_SC), ]) %>% as.vector()),
                   annotation_col = as.data.frame(colData(ddsHT_SC)[,c("type","CTVT_clade", "Origin")]),
                   annotation_row = immune_df[immune_df$geneID %in% rownames(ddsHT_SC), c("cell_type", "cell") ],
                   annotation_colors = row_cols,
                   main = paste0("Heatmap of immune cell related genes, NOT scaled by gene"))


```


