---
title: "03_Deconvolution_CTVT_cibersort"
output: html_document
---

# 0 Environment
## 0.1 Packages
### 0.1.1 Installs
```{r echo=F, message=F, warning=FALSE}
# devtools::install_github("Moonerss/CIBERSORT")
```
### 0.1.1 Library
```{r echo=F, message=F, warning=FALSE}
set.seed(123)
# library(CIBERSORT)
library(tidyverse)
library(SingleCellExperiment)
library(openxlsx)
library(dplyr)
library(Seurat)
# BiocManager::install("orthogene")
```

## 0.2 Paths
```{r echo=F, message=FALSE, warning=FALSE}
fprefix <- "03_Deconvolution_CTVT_cibersort"
wDir <- paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/")
dataDir <- paste0(wDir, "/031_Data/")
resDir <- paste0(wDir, "/033_Results/", fprefix, "/")
plotDir <- paste0(wDir, "/034_Plots/", fprefix, "/")

dir.create(resDir)
dir.create(plotDir)
```

### 0.3.3 Metrics
```{r}
# source(paste0(wDir, "/12_Scripts/01_Deconvolution_deconBenchmark_metricFunctions.R"))
```

# 1.0 Inputs
## 1.1 Candidate marker genes 
```{r}
masterMarkerGenes <- read.xlsx("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_masterMarkerGenes.xlsx")

geneDict <- masterMarkerGenes %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

uniqueGeneDict <- masterMarkerGenes %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

adrianMerged <- read.xlsx(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_deltaEtEh.xlsx")) %>%
  rename("GeneID"="ID")

# lmm24
host <- adrianMerged %>%
  filter(deltaEtEh < 100 & purity_correlation < 0)

# 79 --> either treat all as CTVT specific OR remove and add in marker genes from Adrian's table?
tumor <- adrianMerged %>%
  filter(deltaEtEh > 100 & purity_correlation > 0)

candidateAdrian <- c(host$Symbol, tumor$Symbol)

candidateId <- c(host$Symbol, tumor$Symbol)
```


## 1.1 Reference
Reference - cellxgene dense matrix.
### 1.1.1 Matrix
```{r}
obj <- readRDS(paste0(dataDir, "SCE_concat_CTVT_includingAllGenes_logCorrecDone_vstCorrecDone_ClustDone2_ManAnot.rds"))

objNoBad <- obj[,obj$manual.log.anot.fine!="bad"]

rm(obj)
gc()

rowData <- rowData(objNoBad) %>%
  as.data.frame() %>%
  left_join(adrianMerged) %>%
  drop_na(sampleCount) # not in Adrian's table. 

write.xlsx(rowData, paste0(dataDir, "03_Deconvolution_CTVT_seuratObjNoBad_rowData.xlsx"))

# nrow(rowData) # 14,000. 

# nrow(rowData %>%
#        filter(purity_correlation > 0 & deltaEtEh > 0))

tumor <- rowData %>%
       filter(purity_correlation > 0 & deltaEtEh > 100)

# nrow(tumor) # All 746 --> from which we can filter the snCounts matrix

host <- rowData %>%
       filter(purity_correlation < 0 & deltaEtEh < -100)

# nrow(host) # 332 --> from which we can filter the snCounts matrix. 6 are missing from host that would have otherwise been coming from Adrian's table. 

# snLogCountsFinal <- as.matrix(assay(objNoBad, "logcounts"))

snCountsFinal <- as.matrix(assay(objNoBad, "counts"))

snCountsFinalDf <-  snCountsFinal[unique(geneDict$Symbol),] %>%
  as.data.frame() 

snCountsFinalDfUnique <- snCountsFinal[unique(uniqueGeneDict$Symbol),] %>%
  as.data.frame() 

# snLogCountsFinalDf <- snLogCountsFinal[unique(geneDict$Symbol),] %>%
#   as.data.frame() 

rm(snCountsFinal, snCountsFinalUnique)
gc()

snCountsAdrianDf <- snCountsFinal %>%
  as.data.frame() %>%
  rownames_to_column("Symbol") %>%
  filter(Symbol %in% tumor$Symbol | Symbol %in% host$Symbol) %>%
  column_to_rownames("Symbol")

# Create the cell count matrix: rows = cell types, columns = samples
snCellCountsNoBad <- colData(objNoBad) %>%
  as.data.frame() %>%
  group_by(manual.log.anot.fine,
           Sample) %>%
  summarise(cellCount = n()) %>%
  pivot_wider(names_from = Sample,
              values_from = cellCount,
              values_fill = 0) %>%
  column_to_rownames("manual.log.anot.fine")

snCellPropNoBad <- apply(snCellCountsNoBad, 2, function(x) x/sum(x))

colSums(snCellPropNoBad) # Should all be 1.
```


### 1.1.2 Cell type annotations
Cell type labels - character vector same length as # cells in reference.
Cell state labels - optional, finer granularity character vector. (i.e. myeloid 1 vs myeloid 2). 
Labels should have at least 20 to 50 cells. 
```{r}
snLogCountsFinal_colData <- colData(objNoBad)

snAnno <- snLogCountsFinal_colData$manual.log.anot.fine %>%
  as.data.frame()

print(sort(table(snAnno)))

write_tsv(snAnno, 
          paste0(resDir, fprefix, "_snAnno.tsv"))

rowDataHvgs <- rowData(objNoBad) %>%
  as.data.frame() %>%
  filter(hvgs==TRUE)
```

## 1.3 Bulk
### 1.2.1 Matrix
```{r}
# bulkCountsFinal <- readRDS(paste0(dataDir,
#                                   "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
#   as.data.frame() %>%
#   rownames_to_column("GeneID") %>%
#   filter(GeneID %in% geneDict$GeneID) %>%
#   left_join(geneDict) %>%
#   column_to_rownames("Symbol") %>%
#   select(-GeneID)

bulkCpm <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "CPM")%>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% geneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)

bulkCpmUnique <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "CPM")%>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% uniqueGeneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)
```
### 1.3.2 Metadata
```{r}
bulkCountsFinal_metadata <- read.xlsx(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal_metadata.xlsx"))
```

## 1.4 Write counts to tsv files for CIBERSORT website. 
https://cibersortx.stanford.edu/tutorial.php
```{r}
snCountsExport <- snCountsFinalDf %>%
           t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsExport[1,] <- snAnno$.


snCountsExport <- snCountsExport %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsExport) <- NULL
colnames(snCountsExport) <- NULL

write.table(snCountsExport,
          paste0(resDir, fprefix, "_snCountsFinal_snAnno.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")


snCountsUniqueExport <- snCountsFinalDfUnique %>%
           t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsUniqueExport[1,] <- snAnno$.


snCountsUniqueExport <- snCountsUniqueExport %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsUniqueExport) <- NULL
colnames(snCountsUniqueExport) <- NULL

write.table(snCountsUniqueExport,
          paste0(resDir, fprefix, "_snCountsFinalUnique_snAnno.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")

snCountsExportCurated <- snCountsFinalDf %>%
           t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsExportCurated[1,] <- snAnno$.


snCountsExportCurated <-snCountsExportCurated%>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% curatedMarkers$Symbol)

rownames(snCountsExportCurated) <- NULL
colnames(snCountsExportCurated) <- NULL

write.table(snCountsExportCurated,
          paste0(resDir, fprefix, "_snCountsFinalCurated_snAnno.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")


snCountsExportAdrian <- snCountsAdrianDf %>%
           t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsExportAdrian[1,] <- snAnno$.


snCountsExportAdrian <-snCountsExportAdrian%>%
  rownames_to_column("geneSymbol") 

rownames(snCountsExportAdrian) <- NULL
colnames(snCountsExportAdrian) <- NULL

write.table(snCountsExportAdrian,
          paste0(resDir, fprefix, "_snCountsFinalAdrian_snAnno.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")


bulkCountsFinalExport <- bulkCpm %>%
  as.data.frame() %>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% rownames(snLogCountsFinalDf)) %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  t() %>%
  as.data.frame()
  
rownames(bulkCountsFinalExport) <- NULL
colnames(bulkCountsFinalExport) <- NULL

write.table(bulkCountsFinalExport ,
            paste0(resDir, fprefix, "_bulkCpmFinal.txt"), 
              quote=F,
              row.names = F,
              col.names = F,
              sep = "\t")


bulkCountsFinalUniqueExport <- bulkCpmUnique %>%
  as.data.frame() %>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% snCountsUniqueExport[,1]) %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  t() %>%
  as.data.frame()
  
rownames(bulkCountsFinalUniqueExport) <- NULL
colnames(bulkCountsFinalUniqueExport) <- NULL

write.table(bulkCountsFinalUniqueExport ,
            paste0(resDir, fprefix, "_bulkCpmFinalUnique.txt"), 
              quote=F,
              row.names = F,
              col.names = F,
              sep = "\t")


```


# 2.0 Inputs with purity
## 2.1 Candidate marker genes 
```{r}
masterMarkerGenes <- read.xlsx("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_masterMarkerGenes.xlsx")

geneDict <- masterMarkerGenes %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

uniqueGeneDict <- masterMarkerGenes %>%
  filter(Unique=="Unique") %>%
  select(Symbol, GeneID) %>%
  # Take distinct as marker genes may overlap clusters.
  distinct()

exampleData <- read.table("~/Downloads/Single_Cell_RNA-Seq_Melanoma_SuppFig_3b-d/scRNA-Seq_reference_melanoma_Tirosh_SuppFig_3b-d.txt", 
                          sep = "\t")[1:5,]
```


## 2.1 Reference
Reference - cellxgene dense matrix.
### 2.1.1 Matrix
```{r}
# objNoBad <- obj[,obj$manual.log.anot.fine!="bad"]

colData <- colData(objNoBad)

snCountsFinal <- as.matrix(assay(objNoBad, "counts"))

snCountsFinalDf <-  snCountsFinal[unique(geneDict$Symbol),] %>%
  as.data.frame() 
```


### 2.1.2 Cell type annotations
Cell type labels - character vector same length as # cells in reference.
Cell state labels - optional, finer granularity character vector. (i.e. myeloid 1 vs myeloid 2). 
Labels should have at least 20 to 50 cells. 
```{r}
snLogCountsFinal_colData <- colData(objNoBad)

snAnno <- snLogCountsFinal_colData$manual.log.anot.fine %>%
  as.data.frame()

print(sort(table(snAnno)))

write_tsv(snAnno, 
          paste0(resDir, fprefix, "_snAnno.tsv"))


rowDataHvgs <- rowData(objNoBad) %>%
  as.data.frame() %>%
  filter(hvgs==TRUE)
```

## 2.3 Bulk
### 2.2.1 Matrix
```{r}
bulkCountsFinal <- readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")) %>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% geneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)



bulkCpm <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "CPM")%>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% geneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)


bulkCpmUnique <- DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "CPM")%>%
  as.data.frame() %>%
  rownames_to_column("GeneID") %>%
  filter(GeneID %in% uniqueGeneDict$GeneID) %>%
  left_join(geneDict) %>%
  column_to_rownames("Symbol") %>%
  select(-GeneID)

gtf <- rtracklayer::import('~/gitClones/02_polyARibo/021_Data/Canis_lupus_familiaris.CanFam3.1.104.gtf') %>%
  as.data.frame() %>%
  mutate(geneLength = end-start) %>%
  filter(gene_id %in% rownames(bulkTpm))

bulkTpm <- readRDS(paste0(dataDir,
                       "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds"))

DGEobj.utils::convertCounts(readRDS(paste0(dataDir,
                                  "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal.rds")), 
                                       unit = "TPM",
                                  geneLength = gtf$geneLength)%>%
  as.data.frame()
```
### 2.3.2 Metadata
```{r}
bulkCountsFinal_metadata <- read.xlsx(paste0(dataDir, "01_Deconvolution_matchedBulkMatrixCTVT_bulkCountsFinal_metadata.xlsx"))
```

## 2.4 Write counts to tsv files for CIBERSORT website. 
https://cibersortx.stanford.edu/tutorial.php
```{r}
snLogCountsExport <- snLogCountsFinalDf %>%
            t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snLogCountsExport[1,] <- snAnno$.


snLogCountsExport <- snLogCountsExport %>%
  rownames_to_column("geneSymbol") 

write.table(snLogCountsExport, 
          paste0(resDir, fprefix, "_snLogCountsFinal_snAnno.txt"), 
          # quote=F,
          col.names = F,
          row.names = F, 
          sep = "\t")

test <- read.table(paste0(resDir, fprefix, "_snLogCountsFinal_snAnno.txt"), 
                          sep = "\t")[1:5,]

snCountsExport <- snCountsFinalDf %>%
           t() %>%
            as.data.frame() %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsExport[1,] <- snAnno$.

snCountsExport <- snCountsExport %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsExport) <- NULL
colnames(snCountsExport) <- NULL

write.table(snCountsExport,
          paste0(resDir, fprefix, "_snCountsFinal_snAnno.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")


bulkCountsFinalExport <- bulkCpm %>%
  as.data.frame() %>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% rownames(snLogCountsFinalDf)) %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  t() %>%
  as.data.frame()
  
rownames(bulkCountsFinalExport) <- NULL
colnames(bulkCountsFinalExport) <- NULL

write.table(bulkCountsFinalExport ,
            paste0(resDir, fprefix, "_bulkCpmFinal.txt"), 
              quote=F,
              row.names = F,
              col.names = F,
              sep = "\t")

test <- read.table(paste0(resDir, fprefix, "_bulkCpmFinal.txt"), 
                          sep = "\t")[1:5,]
```

# 2 Results
## 2.1 Process marker gene set
Read in scaled GEP file. 
Trying to reduce the number of CAF.Endo + Myeloid markers. 
```{r}
scaledGep <- read.xlsx(paste0(resDir,
                              "CIBERSORTx_Job15_output/CIBERSORTxGEP_Job15_GEPs_Unfiltered_MeanCentered.xlsx"))

topCAFEndo <- masterMarkerGenes %>%
  filter(cluster == "CAF.Endo") %>%
  filter(Unique=="Unique") %>%
  arrange(deltaEtEh) 

# scaledGep %>%
#   filter(GeneSymbol %in% topCAFEndo[1:10,]$Symbol)

scaledGepCAFEndo <- scaledGep %>%
  arrange(-CAF)

scaledGepCAFEndoTop10 <- scaledGepCAFEndo[1:11, "GeneSymbol"]

topMyeloid <- masterMarkerGenes %>%
  filter(cluster == "Myeloid") %>%
  filter(Unique=="Unique") %>%
  arrange(deltaEtEh) 

# scaledGep %>%
  # filter(GeneSymbol %in% topMyeloid[1:10,]$Symbol)

scaledGepMyeloid <- scaledGep %>%
  arrange(-Myeloid)

scaledGepMyeloidTop10 <- scaledGepMyeloid[1:21, "GeneSymbol"]

scaledGepTcells <- scaledGep %>%
  arrange(-Tcells)

scaledGepTcellsTop10 <- scaledGepTcells[1:21, "GeneSymbol"]

scaledGepBcells <- scaledGep %>%
  arrange(-Bcells)

scaledGepBcellsTop10 <- scaledGepBcells[1:21, "GeneSymbol"]

scaledGepCTVT <- scaledGep %>%
  arrange(-CTVT)

scaledGepCTVTTop10 <- scaledGepCTVT[1:21, "GeneSymbol"]

updatedGepGenes <- c(scaledGepCAFEndoTop10 , 
                           scaledGepMyeloidTop10 , 
                           scaledGepTcellsTop10, 
                           scaledGepBcellsTop10, 
                           scaledGepCTVTTop10)

```

## 2.2 Write updated signature matrix to tsv files for CIBERSORT website. 
https://cibersortx.stanford.edu/tutorial.php
```{r}
# snLogCountsExport <- snLogCountsFinalDf %>%
#             t() %>%
#             as.data.frame() %>%
#             # Not actually GeneSymbol, but for matching format. 
#             rownames_to_column("geneSymbol") %>%
#             filter(geneSymbol %in% updatedGepGenes) %>%
#             t() %>%
#             as.data.frame()
# 
# snLogCountsExport[1,] <- snAnno$.
# 
# 
# snLogCountsExport <- snLogCountsExport %>%
#   rownames_to_column("geneSymbol") 
# 
# write.table(snLogCountsExport, 
#           paste0(resDir, fprefix, "_snLogCountsFinal_snAnno.txt"), 
#           # quote=F,
#           col.names = F,
#           row.names = F, 
#           sep = "\t")
# 
# test <- read.table(paste0(resDir, fprefix, "_snLogCountsFinal_snAnno.txt"), 
#                           sep = "\t")[1:5,]

snCountsExport <- snCountsFinalDf %>%
           t() %>%
            as.data.frame() %>%
            select(updatedGepGenes) %>%
            # Not actually GeneSymbol, but for matching format. 
            rownames_to_column("geneSymbol") %>%
            
            t() %>%
            as.data.frame()

snCountsExport[1,] <- snAnno$.


snCountsExport <- snCountsExport %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsExport) <- NULL
colnames(snCountsExport) <- NULL

write.table(snCountsExport,
          paste0(resDir, fprefix, "_snCountsFinal_snAnno_gep20.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")

test <- read.table(paste0(resDir, 
                          fprefix, 
                          "_snCountsFinal_snAnno_gep20.txt"), 
                          sep = "\t")


test <- read.table(paste0(resDir, fprefix, "_bulkCpmFinal_gep.txt"), 
                          sep = "\t")
```

# 3 Downsample
## 3.1 Convert to Seurat
- Downsample to 500 cells of each cell type
- Use ID's to filter snCountsFinal and snLogCountsFinal
- Export HVGs subset AND candidate marker genes from Adrian's table+FAM results above. 
https://github.com/dyammons/Canine_Leukocyte_scRNA

```{r}
#randomly downsample the subset data to obtain equal number of cells in each celltype
set.seed(123)


assay(objNoBad, "counts") <- as.matrix(assay(objNoBad, "counts"))
assay(objNoBad, "logcounts") <- as.matrix(assay(objNoBad, "logcounts"))
seuratObjNoBad <- as.Seurat(objNoBad, counts = "counts", data = "logcounts")

Idents(seuratObjNoBad) <- "manual.log.anot.fine"

seuratObjNoBadDs <- subset(seuratObjNoBad, downsample = 500)

snAnnoDs <- seuratObjNoBadDs$manual.log.anot.fine

# > table(seuratObjNoBadDs$manual.log.anot.fine)
# 
#   Bcells CAF.Endo     CTVT  Myeloid   Plasma   Tcells 
#      500      500      500      500      310      500 

snCountsFinal <- as.matrix(assay(objNoBad, "counts"))
```

## 3.2 HVGs
### 3.2.1 Counts
```{r}
head(colnames(snCountsFinal))

snCountsFinalDs <- snCountsFinal %>%
  as.data.frame() %>%
  select(colnames(seuratObjNoBadDs)) %>%
  rownames_to_column("genes") %>%
  filter(genes %in% unique(rowDataHvgs$Symbol)) %>%
  column_to_rownames("genes") %>%
  as.matrix()

snCountsFinalDfDsHvgsExport <-  snCountsFinalDs %>%
  as.data.frame() %>%
           t() %>%
            as.data.frame() %>%
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsFinalDfDsHvgsExport[1,] <- snAnnoDs


snCountsFinalDfDsHvgsExport <- snCountsFinalDfDsHvgsExport %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsFinalDfDsHvgsExport) <- NULL
colnames(snCountsFinalDfDsHvgsExport) <- NULL

write.table(snCountsFinalDfDsHvgsExport,
          paste0(resDir, fprefix, "_snCountsFinalDfDsHvgsExport.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")


bulkCountsFinalHvgsExport <- bulkCpm %>%
  as.data.frame() %>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% unique(rowDataHvgs$Symbol)) %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  t() %>%
  as.data.frame()
  
rownames(bulkCountsFinalHvgsExport) <- NULL
colnames(bulkCountsFinalHvgsExport) <- NULL

write.table(bulkCountsFinalHvgsExport ,
            paste0(resDir, fprefix, "_bulkCpmFinalHvgs.txt"), 
              quote=F,
              row.names = F,
              col.names = F,
              sep = "\t")


```

## 3.3 Candidate marker genes 
### 3.3.1 Counts
```{r}
snCountsFinalDs <- snCountsFinal %>%
  as.data.frame() %>%
  select(colnames(seuratObjNoBadDs)) %>%
  rownames_to_column("genes") %>%
  filter(genes %in% unique(geneDict$Symbol)) %>%
  column_to_rownames("genes") %>%
  as.matrix()

snCountsFinalDfDsq1q4Export <-  snCountsFinalDs %>%
  as.data.frame() %>%
           t() %>%
            as.data.frame() %>%
            rownames_to_column("geneSymbol") %>%
            t() %>%
            as.data.frame()

snCountsFinalDfDsq1q4Export[1,] <- snAnnoDs

snCountsFinalDfDsq1q4Export <- snCountsFinalDfDsq1q4Export %>%
  rownames_to_column("geneSymbol") 

rownames(snCountsFinalDfDsq1q4Export) <- NULL
colnames(snCountsFinalDfDsq1q4Export) <- NULL

write.table(snCountsFinalDfDsq1q4Export,
          paste0(resDir,
                 fprefix, 
                 "_snCountsFinalDfDsq1q4Export.txt"),
          quote=F,
          col.names = F,
          row.names = F,
          sep = "\t")

bulkCountsFinalq1q4Export <- bulkCpm %>%
  as.data.frame() %>%
  rownames_to_column("geneSymbol") %>%
  filter(geneSymbol %in% unique(geneDict$Symbol)) %>%
  t()%>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  t() %>%
  as.data.frame()
  
rownames(bulkCountsFinalq1q4Export) <- NULL
colnames(bulkCountsFinalq1q4Export) <- NULL

write.table(bulkCountsFinalq1q4Export ,
            paste0(resDir, fprefix, "_bulkCpmFinalq1q4.txt"), 
              quote=F,
              row.names = F,
              col.names = F,
              sep = "\t")


```

# 5 Visualizations
## 5.1 Set up 
### 5.1.1 snRNAseq
```{r}
# Step 2: Define a color mapping
# Step 2: Define a color mapping
celltype_colors <- c(
  "CTVT" = "#BCD0C7",
  "CAF.Endo" = "#FDDFA4FF",
  "Myeloid" = "#F5FDC6",
  "Plasma" = "#B4B9E0FF",
  "Tcells" = "#D56AA0",
  "Bcells" = "#A8D5E2"
  
)

groundTruth <- snCellPropNoBad %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion") %>%
  mutate(
         Sample = paste0(Sample, " - GT"))

# matchedSamples <- unique(objNoBad$Sample)

# snCountsFinal_colData <- colData(objNoBad)
# 
# # Create the cell count matrix: rows = cell types, columns = samples
# snCellCounts <- snCountsFinal_colData %>%
#   as.data.frame() %>%
#   dplyr::group_by(manual.log.anot.fine, 
#            Sample) %>%
#   summarise(cellCount = n()) %>%
#   pivot_wider(names_from = Sample, 
#               values_from = cellCount, 
#               values_fill = 0) %>%
#   column_to_rownames("manual.log.anot.fine") %>%
#   as.matrix()
# 
# snCellProp <- prop.table(snCellCounts,2) %>%
#   as.data.frame()

# truePurities <- snCellProp  
# 
# truePurities$`3145T11a_n` <- truePurities[,"3145T11a"]
# 
# truePurities$`3145T1c_n` <- truePurities[,"3145T1c"]
# truePurities$`3145T1c_p` <- truePurities[,"3145T1c"]
# 
# truePurities$`3203T1A2c_2` <- truePurities[,"3203T1A2c"]
# truePurities$`3203T1A2c_n` <- truePurities[,"3203T1A2c"]
# 
# truePurities$`3302T1b_n` <- truePurities[,"3302T1b"]
# truePurities$`3302T1b_p` <- truePurities[,"3302T1b"]
# 
# truePurities$`3621T_n` <- truePurities[,"3621T"]
# 
# truePuritiesPlot <- truePurities %>%
#   t() %>%
#   as.data.frame() %>%
#   rownames_to_column("Sample") %>%
#   pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion")

```
### 5.1.2 Bulk
```{r}

```


## 5.1 Min Expr 0, Min Genes 1
### 5.1.1 Non-unique
#### 5.1.1.1 Predictions
```{r}
cibersort <- read_csv(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr0_minGenes1/CIBERSORTx_Job51_Results.csv")) %>%
  column_to_rownames("Mixture") %>%
  select(-c("P-value", "Correlation", "RMSE")) %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion") %>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))

cibersort$Cluster <- str_replace(cibersort$Cluster,
                                 "CAF$",
                                 "CAF.Endo")

cibersort$Cluster <- factor(cibersort$Cluster, levels = rev(c("CTVT", 
                                                  "Myeloid", 
                                                  "Bcells",
                                                  "CAF.Endo",
                                                  "Tcells",
                                                  "Plasma")))

# Step 3: Plot
cibersortPlot <- ggplot(
    cibersort,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  ggtitle("minExpr0_minGenes1")

pdf(paste0(plotDir, "/bar/", fprefix, "_cibersortRes_minExpr0_minGenes1.pdf"),
    height=4, width=7)
print(cibersortPlot)
dev.off()
```
### 5.1.2 Unique
```{r}
cibersort <- read_csv(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr0_minGenes1_unique/CIBERSORTx_Job54_Results.csv")) %>%
  column_to_rownames("Mixture") %>%
  select(-c("P-value", "Correlation", "RMSE")) %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion")%>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))

cibersort$Cluster <- str_replace(cibersort$Cluster,
                                 "CAF$",
                                 "CAF.Endo")

cibersort$Cluster <- factor(cibersort$Cluster, levels = rev(c("CTVT", 
                                                  "Myeloid", 
                                                  "Bcells",
                                                  "CAF.Endo",
                                                  "Tcells",
                                                  "Plasma")))

# Step 3: Plot
cibersortPlot <- ggplot(
    cibersort,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))+
  ggtitle("Unique_minExpr0_minGenes1")


pdf(paste0(plotDir, "/bar/", fprefix, "_cibersortResUnique_minExpr0_minGenes1.pdf"),
    height=4, width=7)
print(cibersortPlot)
dev.off()

```


## 5.2 Min Expr 10, Min Genes 2 to 10
### 5.2.1 Non-unique
```{r}
cibersort <- read_csv(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr10_minGenes2to10/CIBERSORTx_Job43_Results.csv")) %>%
  column_to_rownames("Mixture") %>%
  select(-c("P-value", "Correlation", "RMSE")) %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion")%>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))

cibersort$Cluster <- str_replace(cibersort$Cluster,
                                 "CAF$",
                                 "CAF.Endo")

cibersort$Cluster <- factor(cibersort$Cluster, levels = rev(c("CTVT", 
                                                  "Myeloid", 
                                                  "Bcells",
                                                  "CAF.Endo",
                                                  "Tcells",
                                                  "Plasma")))

# Step 3: Plot
cibersortPlot <- ggplot(
    cibersort,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("minExpr10 minGenes2to10")

pdf(paste0(plotDir, "/bar/", fprefix, "_cibersortRes_minExpr10_minGenes2to10.pdf"),
    height=4, width=7)
print(cibersortPlot)
dev.off()
```

## 5.3 Min Expr 10, Min Genes 4 to 6
### 5.3.1 Non-unique
```{r}
cibersort <- read_csv(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr10_minGenes4to6/CIBERSORTx_Job45_Results.csv")) %>%
  column_to_rownames("Mixture") %>%
  select(-c("P-value", "Correlation", "RMSE")) %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion")%>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))

cibersort$Cluster <- str_replace(cibersort$Cluster,
                                 "CAF$",
                                 "CAF.Endo")

cibersort$Cluster <- factor(cibersort$Cluster, levels = rev(c("CTVT", 
                                                  "Myeloid", 
                                                  "Bcells",
                                                  "CAF.Endo",
                                                  "Tcells",
                                                  "Plasma")))

# Step 3: Plot
cibersortPlot <- ggplot(
    cibersort,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("minExpr10 minGenes4to6")

pdf(paste0(plotDir, "/bar/", fprefix, "_cibersortRes_minExpr10_minGenes4to6.pdf"),
    height=4, width=7)
print(cibersortPlot)
dev.off()
```



## 5.3 Min Expr 50, Min Genes 2 to 10
### 5.3.1 Non-unique
```{r}
cibersort <- read_csv(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr50_minGenes2to10/CIBERSORTx_Job47_Results.csv")) %>%
  column_to_rownames("Mixture") %>%
  select(-c("P-value", "Correlation", "RMSE")) %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  pivot_longer(cols = -c("Sample"), names_to = "Cluster", values_to = "Proportion")%>%
  rbind(groundTruth) %>%
    mutate(Type = ifelse(grepl("- GT$", Sample), 
                         "Ground truth",
                         "Estimated"))

cibersort$Cluster <- str_replace(cibersort$Cluster,
                                 "CAF$",
                                 "CAF.Endo")

cibersort$Cluster <- factor(cibersort$Cluster, levels = rev(c("CTVT", 
                                                  "Myeloid", 
                                                  "Bcells",
                                                  "CAF.Endo",
                                                  "Tcells",
                                                  "Plasma")))

# Step 3: Plot
cibersortPlot <- ggplot(
    cibersort,
    aes(
      x = Sample,
      y = Proportion,
      fill = Cluster,
      pattern = Type  # TRUE/FALSE
    )
  ) +
  geom_bar_pattern(
    stat = "identity",
    pattern_fill = "whitesmoke",           # color of stripes
    pattern_angle = 45,               # diagonal stripes
    pattern_density = 0.01,            # small + widely spaced
    pattern_spacing = 0.01,           # space between stripes
    pattern_key_scale_factor = 0.5, 
    pattern_alpha = 0.4 # smaller legend pattern
  ) +
  scale_pattern_manual(values = c("Estimated" = "none", "Ground truth" = "stripe")) +
  scale_fill_manual(values = celltype_colors) +
  theme_minimal() +
  ylab("Estimated Proportion") +
  xlab("Sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggtitle("minExpr50 minGenes2to10")

pdf(paste0(plotDir, "/bar/", fprefix, "_cibersortRes_minExpr50_minGenes2to10.pdf"),
    height=4, width=7)
print(cibersortPlot)
dev.off()

```

## 5.6 Feature plots with sig. genes
```{r}
assay(objNoBad, "counts") <- as.matrix(assay(objNoBad, "counts"))
assay(objNoBad, "logcounts") <- as.matrix(assay(objNoBad, "logcounts"))
seuratobjNoBad <- as.Seurat(objNoBad, counts = "counts", data = "logcounts")
seuratobjNoBad
Idents(seuratobjNoBad) <- "manual.log.anot.fine"
seuratobjNoBadNoBad <- subset(seuratobjNoBad, idents = "bad", invert=TRUE)
table(Idents(seuratobjNoBadNoBad))
```

### 5.6.1 Min Expr 50, Min Genes 2 to 10
```{r}
sigGenes <- read.table(paste0(resDir, "/snCountsFinalq1q4/sigMatrix_minExpr50_minGenes2to10/CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_phenoclasses.CIBERSORTx_Job46_03_Deconvolution_CTVT_cibersort_snCountsFinal_snAnno_inferred_refsample.bm.K999.txt"), header = T) %>%
  as.data.frame()

sigGenes$NAME <- str_replace(sigGenes$NAME, "DLA.79", "DLA-79")

for(n in sigGenes$NAME){
  
  plot <- FeaturePlot(seuratobjNoBad, n, reduction="TSNE_log_corrected")
  
  pdf(paste0(plotDir, "feature/", fprefix, "_minExpr50_minGenes2to10_", n, ".pdf"),
      height=4,
        width=4)
  print(plot)
  dev.off()
  
}
```


# 6 LMM 22
## 6.1 Read signature matrix
TPM?
Would need to convert HLA --> DLA, and potentially other dog genes?
```{r}
lmm22 <- read_csv(paste0(dataDir, "LM22.csv")) 
lmm22$`Gene symbol` <- str_replace(lmm22$`Gene symbol`, "HLA-", "DLA-")
```

## 6.2 Cross-reference with Adrian's table
### 6.2.1 Read in table
The problem with LM22 is that it is PBMCs --> so no stromal cells.  
Potentially add these as well?
```{r}
adrianMerged <- read.xlsx(paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/033_Results/03_Deconvolution_CTVT_geneCorrelations/03_Deconvolution_CTVT_geneCorrelations_deltaEtEh.xlsx"))

# 188 genes of 547 in the sig matrix. 
host <- adrianMerged %>%
  filter(deltaEtEh < 0 & purity_correlation < 0) %>%
  filter(Symbol %in% lmm22$`Gene symbol`)

# 79 genes --> either treat all as CTVT specific and remove from other cell types OR remove and add in marker genes from Adrian's table? Useful to know what these are most similar to. 
tumor <- adrianMerged %>%
  filter(deltaEtEh > 0 & purity_correlation > 0) %>%
  filter(Symbol %in% lmm22$`Gene symbol`)
```

### 6.2.2 Map orthologs from Adrian's table to human genes
Ex. HLA --> DLA?
```{r}
# library(biomaRt)
# mart = useMart("ensembl")
# listDatasets(mart)
# dog <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "clfamiliaris_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
# dog <- useEnsembl(biomart = "ensembl", dataset = "clfamiliaris_gene_ensembl", mirror = "useast")
# 
# human <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
# human <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
# 
# # Mapping dog gene IDs to human orthologs
# # genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = x , mart = human, attributesL = c("mgi_symbol"), martL = mouse, uniqueRows=T)
# 
# 
# mapping <- getLDS(attributes = c("ensembl_gene_id", "external_gene_name"),
#                   filters = "ensembl_gene_id",
#                   # values = c("ENSCAFG00000012345", "ENSCAFG00000056789"),
#                   mart = dog,
#                   attributesL = c("ensembl_gene_id", "external_gene_name"),
#                   martL = human)
# 
# adrianCheck <- adrianMerged %>% 
#   
# library(orthogene)
# method <- "babelgene"  
# gene_df <- orthogene::convert_orthologs(gene_df = adrianMerged,
#                                         gene_input = "GeneID", 
#                                         gene_output = "columns", 
#                                         input_species = "dog",
#                                         output_species = "human",
#                                         non121_strategy = "drop_both_species",
#                                         method = method) 

dog <- readr::read_tsv("~/Downloads/human_dog_hcop_fifteen_column.txt") 

length(intersect(dog$human_symbol, lmm22$`Gene symbol`)) # 493 / 457 in LMM22
length(intersect(dog$human_symbol, adrianMerged$Symbol)) # 14415 / 17,243 in Adrian's table
length(intersect(lmm22$`Gene symbol`, adrianMerged$Symbol)) # 384 overlapping, just not in q1 or q4 with thresholds???
```

### 6.2.3 Add CTVT column, using genes that intersect in tumor
- What would the baseline expression need to be?
- Could we use the eH from Adrian's table? --> I think this would make more sense. (maybe also in TPM)
- We could also use Adrian's table to select genes we know are stromal genes? --> maybe that's a first idea. 
#### 6.2.3.1 Binary using LM22
For host cells types, make expression 0, for tumor genes, use expression from LM22?
```{r}
lm22Edit <- lm22 %>% 
  mutate(CTVT = )

```

#### 6.2.3.2 Use eT and eH?
For host genes, make expression 0, for tumor genes, use expression from Adrian's table.
```{r}
ctvt <- 

```
# 7 scAtomic Reference
## 7.1 Data
```{r}
# scAtomic <- readRDS("~/Downloads/harmonized_results_matrix_and_metadata_all_patients_all_cells.RDS")
# metaScAtomic <- colData(scAtomic)
```

# 8 Kassandra
## 8.1 Data
If it is not possible to use the snRNAseq ref, we need to use the bulk.
- Simulate bulk expression?
- Need to map genes to DLA as well?
### 8.1.1 Lab expression
```{r}
samplesExpression <- read_tsv()
labExpression <- read_tsv("~/Downloads/laboratory_data_expressions.tsv")
labExpressionMeta <- read_tsv("~/Downloads/laboratory_data_annotation.tsv")

labExpression$`Gene` <- str_replace(labExpression$Gene, "HLA-", "DLA-")

# Would potentially want to remove these.
# 293
checkLabExpressionTumorExport <- labExpression %>%
  filter(Gene %in% tumor$Symbol) %>%
  column_to_rownames("Gene") 

# 655
checkLabExpressionHostExport <- labExpression %>%
  filter(Gene %in% host$Symbol) %>%
  column_to_rownames("Gene") 

metaHost <- labExpressionMeta %>%
  rename("...1" = "Sample",
         "Cell_type"="cellType") %>%
  filter(Sample %in% colnames(checkLabExpressionHostExport))

metaHostTable <- table(metaHost$cellType)
```

