---
title: "03_Deconvolution_CTVT_Junho2024"
output: html_document
---

# 0 Environment
## 0.1 Packages
### 0.1.1 Installs
### 0.1.1 Library
```{r echo=F, message=F, warning=FALSE}
set.seed(123)
library(tidyverse)
library(openxlsx)
library(SingleCellExperiment)
setClassUnion("ExpData", c("matrix", "SummarizedExperiment"))
library(DGEobj.utils)
library(optparse)
library(dplyr)
library(Biobase)
library(ggpattern)
library(ggnewscale)
library(Seurat)
library(zellkonverter)
library(dplyr)
```

## 0.2 Paths
```{r echo=F, message=FALSE, warning=FALSE}
wDir <- paste0("~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/")
fprefix <- "03_Deconvolution_CTVT_GSE210347"
dataDir <- paste0(wDir, "/031_Data/GSE210347/")
resDir <- paste0(wDir, "/033_Results/", fprefix, "/")
plotDir <- paste0(wDir, "/034_Plots/", fprefix, "/")

dir.create(resDir)
dir.create(plotDir)
```

# 1 Raw data
## 1.1 Raw data
### 1.1.1 Read in data from Luo TME atlas
"...After filtering, 855,271 high-quality cells were preserved for subsequent analyses."
Cells have already been processed and integrated. 
Testing on 5 cell types only (otherwise object is very large!)
148 primary tumor, 53 adjacent, and 25 normal samples from 164
donors enrolled in 12 studies. 
Idea is to: check all cells for cnv --> remove all tumor cells. 
Alternatively, remove primary tumor and normal samples. 
!!! Have done what is shown below !!!
Next, unsupervised clustering generated a total of 34 TME-related clusters (c1-c34, 569,759 cells),
which were separated from cancerous/normal epithelial cells (Fig. 1c and d).
```{r}
temp <- readRDS(paste0(dataDir, "/GSE210347_counts.Rds"))

meta <-read.delim(paste0(dataDir, "/GSE210347_meta.txt")) %>%
  column_to_rownames("cellname")

sampleInfo <-read.delim(paste0(dataDir, "/GSE210347_sample_information.txt"))

temp <- CreateSeuratObject(counts = temp,
                           meta = meta)
print(temp)

# temp[["percent.mt"]] <- PercentageFeatureSet(temp, pattern = "^MT-")

table(temp$celltype)

Idents(temp) <- "celltype"

subset <- subset(temp, idents= c("Endothelium","Fibroblast", "Lymphocyte", "Myeloid", "Plasma"))

rm(temp)
gc()

print(subset) # 569919 cells. 

saveRDS(subset, paste0(dataDir, "GSE210347_subset.rds"))
```
## 1.2 Pre-process
### 1.2.1 MT
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
temp[["percent.mt"]] <- PercentageFeatureSet(temp, pattern = "^MT-")
```

### 1.2.2 Log normalize
Per cell, normalize gene expression by total expression, multiply by scale factor, and log-transform result. 
Stored in "data".
Assumes each cell contains the same number of RNA molecules. 
```{r}
temp <- NormalizeData(temp, normalization.method = "LogNormalize", scale.factor = 10000)
gc()
```

### 1.2.3 Variable features
Genes with high cell-cell variation. Focusing on these genes helps identify biological signal in the dataset. 
FindVariableFeatures() models the mean-variance relationship in single-cell data. 
```{r}
temp <- FindVariableFeatures(temp, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(temp), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(temp)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

gc()
```

### 1.2.4 Scale data 
Shift expression per gene over all cells so mean = 0 and stdev is 1. 
Variable features only will be scaled, unless defined otherwise. 
Can also be used to remove unwanted sources of variation. 
In paper, vars.to.regress = UMI (counts) + percent mt. 
- Takes a while!
- Is vars.to.regress nCount_RNA correct? --> did not do this. Just default percent.mt for now. 
```{r}
temp <- ScaleData(temp, vars.to.regress = c("percent.mt"))
```

```{r}
saveRDS(temp, paste0(dataDir, "GSE210347/1.2.4_GSE210347.rds"))
```

### 1.2.5 Perform PCA on scaled data
Variable features by default. 
In paper, PCA 1:30 used. 
Per PC, can also return the genes with the highest correlation and anticorrelation across single cells in the dataset. 
Using a higher number of PCs, you are more likely to capture rare immune subsets (ex. DC and NK genes)
```{r}
temp <- RunPCA(temp, features = VariableFeatures(object = temp))
ElbowPlot(pbmc)
```

### 1.2.6 Batch correction
RunFastMNN to correct sample-level batch effect
MNN pairs - find pairs between cells of the same type or state across batches (assuming effect is orthogonal)
Take correction vector per pair (can be an average for all correction vectors for all pairs a cell is part of)
This removes variation within each batch (avoiding the "kissing" problem of dense subpopulations in a batch.)
```{r}
obj <- IntegrateLayers(
  object = obj, method = FastMNNIntegration,
  new.reduction = "integrated.mnn",
  verbose = FALSE
)
```

### 1.2.7 Clustering
Use FindNeighbors and FindClusters using the Louvain algorithm. 
Cells are embedded in a graph structure, drawing edges between highly similar cells 
KNN graph first drawn based on PCA, edge weights are refined by shraed overlap in local neighborhood (Jaccard similarity). 
Iteratively group cells together with the Louvain algorithm.

In paper, 34 clusters were found, looking for a resolution that does this. There is a seurat_clusters column in the metadata, but also a cluster column (1-35).

### 1.2.7 Run UMAP
As a non-linear dim reduc method, preserve local distances (ex. two cells in the same cluster) rather than global relationships. 
Do not draw conclusions based on visuals along. 
```{r}

```

### 1.1.2 Read in data from original paper.
Original paper had 48,164 cells. 
In Junho, there are only annotations from 47,116 cells. 

Metadata of original paper: where the tumor came from
Keep only tumor samples!
How to separate tumor cells from host TME cells? Remove epithelial cluster completely?
```{r}
meta <- read.delim(paste0(dataDir, "GSE144236/GSE144236_patient_metadata_new.txt")) %>%
  as.data.frame()

table(meta$tum.norm)
table(meta$patient)
table(meta$level1_celltype)
# counts <- read.delim(paste0(dataDir, "GSE144236/GSE144236_cSCC_counts.txt"))

metaTumor <- meta %>%
  filter(tum.norm=="Tumor")
# table(metaTumor$level1_celltype)
#             ASDC           B Cell             CD1C           CLEC9A Endothelial Cell       Epithelial       Fibroblast               LC              Mac 
#              319              250             3615              719              194            12651              647             2341             2370 
#             MDSC       Melanocyte        Multiplet               NK              PDC            Tcell 
#              347              530              793              112              248             1163 

metaHost <- meta %>%
  filter(tum.norm != "Tumor")
table(metaHost$level1_celltype)
            # ASDC           B Cell             CD1C           CLEC9A Endothelial Cell       Epithelial       Fibroblast               LC              Mac 
            #   26                6             1298              146              258            16179              235             1461              815 
            # MDSC       Melanocyte        Multiplet               NK              PDC            Tcell 
            #  378              266              303               27                6              461 

counts <-  read.delim(paste0(dataDir, "GSE144236/GSE144236_cSCC_counts.txt"))

human <- CreateSeuratObject(counts = counts,
                            meta.data = meta)
```


## 1.3 Mapping genes
Read in file from ~/gitClones/01_Deconvolution/03_Deconvolution_CTVT/032_Scripts/03_Deconvolution_CTVT_customScaden/03_Deconvolution_CTVT_orthologs.R. 
This maps unique genes between the skin and CTVT datasets. 
